{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import pandas as pd\n",
    "import numpy\n",
    "import numpy as np\n",
    "import csv\n",
    "import Bio\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.Seq import Seq\n",
    "import time\n",
    "import numpy as np\n",
    "from Bio.Data.CodonTable import unambiguous_dna_by_id\n",
    "from itertools import chain\n",
    "from Bio.SeqUtils.CodonUsage import CodonsDict\n",
    "from Bio.Data import CodonTable\n",
    "from collections import Counter\n",
    "import re\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from os import chdir\n",
    "from statistics import mean, pstdev\n",
    "from scipy.stats.mstats import gmean\n",
    "from scipy.spatial import distance\n",
    "from scipy.stats.stats import pearsonr \n",
    "from scipy import stats\n",
    "from scipy.stats import ks_2samp\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"seq.csv\")\n",
    "gene_sequence = df['seq']\n",
    "df2 = pd.read_csv(\"max_5_ind.csv\")\n",
    "GFP_seq = 'atgtccaagggtgaagagctatttactggggttgtacccattttggtagaactggacggagatgtaaacggacataaattctctgttagaggtgagggcgaaggcgatgccaccaatggtaaattgactctgaagtttatatgcactacgggtaaattacctgttccttggccaaccctagtaacaactttgacatatggtgttcaatgtttctcaagatacccagaccatatgaaaaggcatgatttctttaaaagtgctatgccagaaggctacgtgcaagagagaactatctcctttaaggatgacggtacgtataaaacacgagcagaagtgaaattcgaaggggatacactagttaatcgcatcgaattaaagggtatagactttaaggaagatggtaatattctcggccataaacttgagtataatttcaactcgcataatgtgtacattacagctgacaaacaaaagaacggaattaaagcgaattttaaaatcaggcacaacgtcgaagatgggtctgttcaacttgccgatcattatcagcaaaacacccctattggtgatggtccagtcttgttacccgataatcactacttaagcacacagtctagattgtcaaaagatccgaatgaaaagcgtgatcacatggttttattggaatttgtcaccgctgcaggaataactcacggaatggacgagctttataagggatcc'\n",
    "mCherry_seq = 'atggtgagcaagggcgaggaggataacatggccatcatcaaggagttcatgcgcttcaaggtgcacatggagggctccgtgaacggccacgagttcgagatcgagggcgagggcgagggccgcccctacgagggcacccagaccgccaagctgaaggtgaccaagggtggccccctgcccttcgcctgggacatcctgtcccctcagttcatgtacggctccaaggcctacgtgaagcaccccgccgacatccccgactacttgaagctgtccttccccgagggcttcaagtgggagcgcgtgatgaacttcgaggacggcggcgtggtgaccgtgacccaggactcctccctgcaggacggcgagttcatctacaaggtgaagctgcgcggcaccaacttcccctccgacggccccgtaatgcagaagaagaccatgggctgggaggcctcctccgagcggatgtaccccgaggacggcgccctgaagggcgagatcaagcagaggctgaagctgaaggacggcggccactacgacgctgaggtcaagaccacctacaaggccaagaagcccgtgcagctgcccggcgcctacaacgtcaacatcaagttggacatcacctcccacaacgaggactacaccatcgtggaacagtacgaacgcgccgagggccgccactccaccggcggcatggacgagctgtacaag'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_seq_2_aa(gene_seq):\n",
    "    gene_seq_aa = Seq(gene_seq)\n",
    "    gene_seq_aa = gene_seq_aa.translate()\n",
    "    gene_seq_aa = str (gene_seq_aa)\n",
    "    return gene_seq_aa\n",
    "\n",
    "def genes_seq_2_aa(genes_seq):\n",
    "    genes_seq_aa = []\n",
    "    for i in range(0,len(genes_seq)):\n",
    "        genes_seq_aa.append(gene_seq_2_aa(genes_seq[i]))\n",
    "    return genes_seq_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#finds the longest common string of S and T \n",
    "def lcs(S,T):\n",
    "    m = len(S)\n",
    "    n = len(T)\n",
    "    counter = [[0]*(n+1) for x in range(m+1)]\n",
    "    longest = 0\n",
    "    lcs_set = set()\n",
    "    for i in range(m):\n",
    "        for j in range(n):\n",
    "            if S[i] == T[j]:\n",
    "                c = counter[i][j] + 1\n",
    "                counter[i+1][j+1] = c\n",
    "                if c > longest:\n",
    "                    lcs_set = set()\n",
    "                    longest = c\n",
    "                    lcs_set.add(S[i-c+1:i+1])\n",
    "                elif c == longest:\n",
    "                    lcs_set.add(S[i-c+1:i+1])\n",
    "    x = list(lcs_set)\n",
    "    if len(x) == 0:\n",
    "        return (0,'')\n",
    "    else:\n",
    "        return (len(x[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns the length of the longets coomon string of the target gene and an essential gene\n",
    "# call this function with: LCW(target_gene,gene_sequence[i])\n",
    "def LCW(target_gene,gene_seq): \n",
    "    return lcs(gene_seq_2_aa(target_gene),gene_seq_2_aa(gene_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# given a sequence the function returns its GC content\n",
    "def gc_content(gene_seq):\n",
    "    x = gene_seq.upper()\n",
    "    c = x.count('C')\n",
    "    g = x.count('G')\n",
    "    gc_total = g+c\n",
    "    gc_content = (gc_total/len(x))\n",
    "    return gc_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the function calculates the gc_content profile in a window of 30 nucletides each time (300 nt window in tot).\n",
    "# if the sequence length is less than 300 the rest of the profile will be filll with the average of the gc content of existing windows \n",
    "# ind -1 for 300 nt from the beginning of the sequenece (for essential genes) or 1 for 300 until the end of the sequenece (for target gene)\n",
    "def gc_content_profile(gene_seq,ind):\n",
    "    gc_content_profile = []\n",
    "    i = 0\n",
    "    if len(gene_seq)>299:\n",
    "        if ind == -1:\n",
    "            wind = gene_seq[0:300]\n",
    "            while i < len(wind)-1:\n",
    "                gc_content_profile.append(gc_content(wind[i:i+30]))\n",
    "                i = i + 30\n",
    "        if ind == 1:\n",
    "            wind = gene_seq[len(gene_seq)-300:]\n",
    "            while i < len(wind)-1:\n",
    "                gc_content_profile.append(gc_content(wind[i:i+30]))\n",
    "                i = i + 30\n",
    "    else:\n",
    "        if ind == -1:\n",
    "            x = math.floor(len(gene_seq)/30)\n",
    "            j = 0\n",
    "            for i in range(0,x):\n",
    "                gc_content_profile.append(gc_content(gene_seq[j:j+30]))\n",
    "                j = j + 30\n",
    "            m = mean(gc_content_profile)\n",
    "            for i in range(x,10):\n",
    "                gc_content_profile.append(m)\n",
    "        if ind == 1:\n",
    "            x = math.floor(len(gene_seq)/30)\n",
    "            j = 0\n",
    "            d = len(gene_seq) - (x*30)\n",
    "            wind = gene_seq[d:]\n",
    "            for i in range(0,x):\n",
    "                gc_content_profile.append(gc_content(wind[j:j+30]))\n",
    "                j = j + 30\n",
    "            m = mean(gc_content_profile)\n",
    "            for i in range(x,10):\n",
    "                gc_content_profile.append(m)        \n",
    "    return gc_content_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculates ditances of gc content profiles from target gene to essential gene, and to its closest gene\n",
    "# returns 6 distances from essential gene and 6 from its closet genes (from the avergae on their profiles)\n",
    "# paramters required: essential gene sequence[i], target gene sequence, df2.iloc[i] (the coresponding row in df2 of the chosen essential gene),all yeast genes sequence\n",
    "# _essen mean a distance from eseential gene\n",
    "# _close mean a distance from eseential gene closest genes\n",
    "\n",
    "def calc_dist_gc(gene_sequence,target_gene,df2,genes_sequence):\n",
    "    \n",
    "    gc_target_gene = np.asarray(gc_content_profile(target_gene,1))\n",
    "    gc_gene = np.asarray(gc_content_profile(gene_sequence,-1))\n",
    "    \n",
    "    euc_dist_gc_essen = distance.euclidean(gc_target_gene,gc_gene)\n",
    "    cart_d_gc_essen = sum(gc_target_gene*gc_gene)\n",
    "    corr_p_gc_essen = pearsonr(gc_target_gene,gc_gene)[0]\n",
    "    corr_s_gc_essen = stats.spearmanr(gc_target_gene,gc_gene)[0]\n",
    "    L1_dist_gc_essen = sum(np.abs(gc_target_gene-gc_gene))\n",
    "    ks_dist_gc_essen = ks_2samp(gc_target_gene,gc_gene)[0]\n",
    "    \n",
    "    max_5_ind_1 = df2['max_5_ind_1']\n",
    "    max_5_ind_2 = df2['max_5_ind_2']\n",
    "    max_5_ind_3 = df2['max_5_ind_3']\n",
    "    max_5_ind_4 = df2['max_5_ind_4']\n",
    "    max_5_ind_5 = df2['max_5_ind_5']\n",
    "    \n",
    "    if max_5_ind_1 == -2:\n",
    "        euc_dist_gc_close = []\n",
    "        corr_p_gc_close = []\n",
    "        corr_s_gc_close = []\n",
    "        ks_dist_gc_close = []\n",
    "        cart_d_gc_close = []\n",
    "        L1_dist_gc_close = []\n",
    "    else:\n",
    "        ind = [max_5_ind_1,max_5_ind_2,max_5_ind_3,max_5_ind_4,max_5_ind_5]\n",
    "        for index in ind:\n",
    "            if index==-2:\n",
    "                ind.remove(index)\n",
    "        close_genes_gc = []\n",
    "        for j in range(0,len(ind)):\n",
    "            index = ind[j]\n",
    "            close_genes_gc.append(gc_content_profile(genes_sequence[index],-1))\n",
    "        m_close_genes_gc = np.mean(close_genes_gc,axis=0)\n",
    "    \n",
    "        euc_dist_gc_close = distance.euclidean(gc_target_gene,m_close_genes_gc)\n",
    "        cart_d_gc_close = sum(gc_target_gene*m_close_genes_gc)\n",
    "        corr_p_gc_close = pearsonr(gc_target_gene,m_close_genes_gc)[0]\n",
    "        corr_s_gc_close = stats.spearmanr(gc_target_gene,m_close_genes_gc)[0]\n",
    "        L1_dist_gc_close = sum(np.abs(gc_target_gene-m_close_genes_gc))\n",
    "        ks_dist_gc_close = ks_2samp(gc_target_gene,m_close_genes_gc)[0]\n",
    "         \n",
    "    return(euc_dist_gc_essen,corr_p_gc_essen,corr_s_gc_essen,ks_dist_gc_essen,cart_d_gc_essen,L1_dist_gc_essen,euc_dist_gc_close,corr_p_gc_close,corr_s_gc_close,ks_dist_gc_close,cart_d_gc_close,L1_dist_gc_close)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def altcodons(codon, table):\n",
    "    \"\"\"List codons that code for the same aminonacid / are also stop. \n",
    "    Given a codon, the output is a list of the alternative (synonymous) codons.\n",
    "\n",
    "    @param codon\n",
    "    @table code table id (1 is standard)\n",
    "    @return list of codons\n",
    "\n",
    "    \"\"\"\n",
    "    tab = unambiguous_dna_by_id[table]\n",
    "\n",
    "    if codon in tab.stop_codons:\n",
    "        return tab.stop_codons\n",
    "\n",
    "    try:\n",
    "        aa = tab.forward_table[codon]\n",
    "    except:\n",
    "        return []\n",
    "\n",
    "    return [k for (k, v) in tab.forward_table.items()\n",
    "            if v == aa]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def degeneration(codon, table):\n",
    "    \"\"\"Determine how many codons code for the same amino acid / are also stop (degeneration degree of the input codon)\n",
    "\n",
    "    @param codon the codon\n",
    "    @param table code table id\n",
    "    @return the number of codons also coding for the amino acid codon codes for\n",
    "\n",
    "    \"\"\"\n",
    "    return len(altcodons(codon, table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_table():\n",
    "    \"\"\"Define the table that maps each amino acid to its synonymus codons (without stop codons).\n",
    "    @return a dictionary, where the keys are amino acids and the values are the codons.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    standard_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n",
    "    reverse_genetic_code = standard_table.back_table.copy()\n",
    "\n",
    "    for aa in reverse_genetic_code.keys():\n",
    "        curr_amino_acid = aa\n",
    "        primary_codon = reverse_genetic_code[aa]\n",
    "        syno_codons = altcodons(primary_codon, 1)\n",
    "        reverse_genetic_code[curr_amino_acid] = syno_codons\n",
    "\n",
    "    reverse_genetic_code.pop(None, None)\n",
    "    return reverse_genetic_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_codons(gene_seq):\n",
    "    error_dict = CodonsDict.copy() #empty dictionary\n",
    "    error_dict = {x:np.NaN for x in error_dict}\n",
    "    counts = CodonsDict.copy()\n",
    "    b = len(gene_seq)%3\n",
    "    for i in range(0,len(gene_seq)-b,3):\n",
    "        codon = gene_seq[i:i+3]\n",
    "        counts[codon] += 1\n",
    "    return counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_relative_freq(gene_seq):\n",
    "    \"\"\"For each codon, the function calculates its relative frequency,\n",
    "    i.e the number of apperances of the codon,\n",
    "    divided by the number of appereances of the amino-acid that the codon encodes.\n",
    "    if the codon is not in the sequence, its relative frequency is set to NaN.\n",
    "    \n",
    "    @param gene_seq the sequence of the gene (not string type!)\n",
    "    @return a dictionary where the keys are codons and the values are the relative frquencies.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    counts = count_codons(gene_seq)\n",
    "    gene_seq\n",
    "    genetic_code = unambiguous_dna_by_id[1].forward_table #keys = codons; values = Amino acids\n",
    "    x = {'TGA':'*','TAA':'*','TAG':'*'}\n",
    "    genetic_code.update(x)\n",
    "    reverse_genetic_code = define_table() #keys = Amino acids, values = codons\n",
    "    protein_seq = Seq(gene_seq).translate() \n",
    "    return {codon: float(counts[codon])/protein_seq.count(genetic_code[codon]) if protein_seq.count(genetic_code[codon])!=0 else 0 \n",
    "            for codon in counts.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# given the output of the function above (def codon_relative_freq(gene_seq)), this function returns a \n",
    "def crf_vec(crf):\n",
    "    d = []\n",
    "    for key,value in crf.items():\n",
    "        d.append(value)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the function calculates the frequency of each amino acid given a sequence\n",
    "def AA_freq(gene_seq):\n",
    "    AA = ['F','L','S','Y','C','W','P','M','Q','H','I','R','K','N','T','V','A','E','G','D','*']\n",
    "    aa_freq = []\n",
    "    for aa in AA:\n",
    "        aa_freq.append((gene_seq.count(aa))/len(gene_seq))\n",
    "    return aa_freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function return the distances from the target gene to essential gene for the profiles of codon relative frequency and amino acid relative frequency\n",
    "# in addition, the distances from the target gene to essential gene closet genes for the profiles of codon relative frequency and amino acid relative frequency\n",
    "# paramters required: essential gene sequence[i], target gene sequence, df2.iloc[i] (the coresponding row in df2 of the chosen essential gene),all yeast genes sequence\n",
    "# _essen means a distance from eseential gene\n",
    "# _close means a distance from eseential gene closest genes\n",
    "# _aa means distance for amino acis profile\n",
    "# _c means distance for codons profile\n",
    "\n",
    "def calc_dis_near_genes(gene_sequence,target_gene,df2,genes_sequence):\n",
    "    \n",
    "    aaf = np.asarray(AA_freq(gene_seq_2_aa(target_gene)))\n",
    "    crf = np.asarray(crf_vec(codon_relative_freq(target_gene.upper())))\n",
    "    gene_seq_aaf = np.asarray(AA_freq(gene_seq_2_aa(gene_sequence)))\n",
    "    gene_seq_crf = np.asarray(crf_vec(codon_relative_freq(gene_sequence)))\n",
    "    \n",
    "    euc_dist_c_essen = distance.euclidean(crf,gene_seq_crf)\n",
    "    cart_d_c_essen = sum(crf*gene_seq_crf)\n",
    "    corr_p_c_essen = pearsonr(crf,gene_seq_crf)[0]\n",
    "    corr_s_c_essen = stats.spearmanr(crf,gene_seq_crf)[0]\n",
    "    L1_dist_c_essen = sum(np.abs(crf-gene_seq_crf))\n",
    "    ks_dist_c_essen = ks_2samp(crf,gene_seq_crf)[0]\n",
    "    \n",
    "    euc_dist_aa_essen = distance.euclidean(aaf,gene_seq_aaf)\n",
    "    cart_d_aa_essen = sum(aaf*gene_seq_aaf)\n",
    "    corr_p_aa_essen = pearsonr(aaf,gene_seq_aaf)[0]\n",
    "    corr_s_aa_essen = stats.spearmanr(aaf,gene_seq_aaf)[0]\n",
    "    L1_dist_aa_essen = sum(np.abs(aaf-gene_seq_aaf))\n",
    "    ks_dist_aa_essen = ks_2samp(aaf,gene_seq_aaf)[0]\n",
    "    \n",
    "    max_5_ind_1 = df2['max_5_ind_1']\n",
    "    max_5_ind_2 = df2['max_5_ind_2']\n",
    "    max_5_ind_3 = df2['max_5_ind_3']\n",
    "    max_5_ind_4 = df2['max_5_ind_4']\n",
    "    max_5_ind_5 = df2['max_5_ind_5']\n",
    "    \n",
    "    if max_5_ind_1 == -2:\n",
    "        \n",
    "        euc_dist_c_close = []\n",
    "        cart_d_c_close = []\n",
    "        corr_p_c_close = []\n",
    "        corr_s_c_close = []\n",
    "        L1_dist_c_close = []\n",
    "        ks_dist_c_close = []\n",
    "    \n",
    "        euc_dist_aa_close = []\n",
    "        cart_d_aa_close = []\n",
    "        corr_p_aa_close = []\n",
    "        corr_s_aa_close = []\n",
    "        L1_dist_aa_close = []\n",
    "        ks_dist_aa_close = []\n",
    "    else:\n",
    "        ind = [max_5_ind_1,max_5_ind_2,max_5_ind_3,max_5_ind_4,max_5_ind_5]\n",
    "        for index in ind:\n",
    "            if index==-2:\n",
    "                ind.remove(index)\n",
    "        close_genes_c = []\n",
    "        close_genes_aa = []\n",
    "        for j in range(0,len(ind)):\n",
    "            index = ind[j]\n",
    "            close_genes_c.append(crf_vec(codon_relative_freq(genes_sequence[index])))\n",
    "            close_genes_aa.append(AA_freq(gene_seq_2_aa(genes_sequence[index])))\n",
    "        m_close_genes_c = np.mean(close_genes_c,axis=0)\n",
    "        m_close_genes_aa = np.mean(close_genes_aa,axis=0)\n",
    "    \n",
    "        euc_dist_c_close = distance.euclidean(crf,m_close_genes_c)\n",
    "        cart_d_c_close = sum(crf*m_close_genes_c)\n",
    "        corr_p_c_close = pearsonr(crf,m_close_genes_c)[0]\n",
    "        corr_s_c_close = stats.spearmanr(crf,m_close_genes_c)[0]\n",
    "        L1_dist_c_close = sum(np.abs(crf-m_close_genes_c))\n",
    "        ks_dist_c_close = ks_2samp(crf,m_close_genes_c)[0]\n",
    "        \n",
    "        euc_dist_aa_close = distance.euclidean(aaf,m_close_genes_aa)\n",
    "        cart_d_aa_close = sum(aaf*m_close_genes_aa)\n",
    "        corr_p_aa_close = pearsonr(aaf,m_close_genes_aa)[0]\n",
    "        corr_s_aa_close = stats.spearmanr(aaf,m_close_genes_aa)[0]\n",
    "        L1_dist_aa_close = sum(np.abs(aaf-m_close_genes_aa))\n",
    "        ks_dist_aa_close = ks_2samp(aaf,m_close_genes_aa)[0]\n",
    "          \n",
    "    return(euc_dist_c_essen,corr_p_c_essen,corr_s_c_essen,ks_dist_c_essen,cart_d_c_essen,L1_dist_c_essen,euc_dist_aa_essen,corr_p_aa_essen,corr_s_aa_essen,ks_dist_aa_essen,cart_d_aa_essen,L1_dist_aa_essen,euc_dist_c_close,corr_p_c_close,corr_s_c_close,ks_dist_c_close,cart_d_c_close,L1_dist_c_close,euc_dist_aa_close,corr_p_aa_close,corr_s_aa_close,ks_dist_aa_close,cart_d_aa_close,L1_dist_aa_close)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the function calculates the folding energy profile in a window of 40 nucletides each time and a shift of 10 nt (150 nt window in tot).\n",
    "# if the sequence length is less than 300 the rest of the profile will be filll with the average of the gc content of existing windows \n",
    "# ind -1 for 300 nt from the beginning of the sequenece (for essential genes) or 1 for 300 until the end of the sequenece (for target gene)\n",
    "def fe_profile(gene_seq,ind):\n",
    "    fe_profile = []\n",
    "    i = 0\n",
    "    j = 0\n",
    "    if len(gene_seq)>150:\n",
    "        if ind == -1:\n",
    "            wind = gene_seq[0:150]\n",
    "            while j < 12:\n",
    "                (ss, mfe) = RNA.fold(wind[i:i+40])\n",
    "                fe_content_profile.append(mfe)\n",
    "                i = i + 10\n",
    "                j = j + 1\n",
    "        if ind == 1:\n",
    "            wind = gene_seq[len(gene_seq)-150:]\n",
    "            while j < 12:\n",
    "                (ss, mfe) = RNA.fold(wind[i:i+40])\n",
    "                fe_content_profile.append(mfe)\n",
    "                i = i + 10\n",
    "                j = j + 1\n",
    "    else:\n",
    "        if ind == -1:\n",
    "            x = (math.floor(len(gene_seq)/10))-3\n",
    "            j = 0\n",
    "            for i in range(0,x):\n",
    "                (ss, mfe) = RNA.fold(wind[j:j+40])\n",
    "                fe_content_profile.append(mfe)\n",
    "                j = j + 10\n",
    "            m = mean(fe_content_profile)\n",
    "            for i in range(x,12):\n",
    "                fe_content_profile.append(m)\n",
    "        if ind == 1:\n",
    "            x = (math.floor(len(gene_seq)/10))-3\n",
    "            j = 0\n",
    "            d = len(gene_seq)%10\n",
    "            wind = gene_seq[d:]\n",
    "            for i in range(0,x):\n",
    "                (ss, mfe) = RNA.fold(wind[j:j+40])\n",
    "                fe_content_profile.append(mfe)\n",
    "                j = j + 10\n",
    "            m = mean(fe_content_profile)\n",
    "            for i in range(x,12):\n",
    "                fe_content_profile.append(m)        \n",
    "    return fe_content_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculates ditances of folding energy profiles from target gene to essential gene, and to its closest gene\n",
    "# returns 6 distances from essential gene and 6 from its closet genes (from the avergae on their profiles)\n",
    "# paramters required: essential gene sequence[i], target gene sequence, df2.iloc[i] (the coresponding row in df2 of the chosen essential gene),all yeast genes sequence\n",
    "# _essen mean a distance from eseential gene\n",
    "# _close mean a distance from eseential gene closest genes\n",
    "\n",
    "def calc_dist_fe(gene_sequence,target_gene,df2,genes_sequence):\n",
    "    \n",
    "    fe_target_gene = np.asarray(fe_profile(target_gene,1))\n",
    "    fe_gene = np.asarray(fe_profile(gene_sequence,-1))\n",
    "    \n",
    "    euc_dist_fe_essen = distance.euclidean(fe_target_gene,fe_gene)\n",
    "    cart_d_fe_essen = sum(fe_target_gene*fe_gene)\n",
    "    corr_p_fe_essen = pearsonr(fe_target_gene,fe_gene)[0]\n",
    "    corr_s_fe_essen = stats.spearmanr(fe_target_gene,fe_gene)[0]\n",
    "    L1_dist_fe_essen = sum(np.abs(fe_target_gene-fe_gene))\n",
    "    ks_dist_fe_essen = ks_2samp(fe_target_gene,fe_gene)[0]\n",
    "    \n",
    "    max_5_ind_1 = df2['max_5_ind_1']\n",
    "    max_5_ind_2 = df2['max_5_ind_2']\n",
    "    max_5_ind_3 = df2['max_5_ind_3']\n",
    "    max_5_ind_4 = df2['max_5_ind_4']\n",
    "    max_5_ind_5 = df2['max_5_ind_5']\n",
    "    \n",
    "    if max_5_ind_1 == -2:\n",
    "        euc_dist_fe_close = []\n",
    "        corr_p_fe_close = []\n",
    "        corr_s_fe_close = []\n",
    "        ks_dist_fe_close = []\n",
    "        cart_d_fe_close = []\n",
    "        L1_dist_fe_close = []\n",
    "    else:\n",
    "        ind = [max_5_ind_1,max_5_ind_2,max_5_ind_3,max_5_ind_4,max_5_ind_5]\n",
    "        for index in ind:\n",
    "            if index==-2:\n",
    "                ind.remove(index)\n",
    "        close_genes_fe = []\n",
    "        for j in range(0,len(ind)):\n",
    "            index = ind[j]\n",
    "            close_genes_fe.append(fe_profile(genes_sequence[index],-1))\n",
    "        m_close_genes_fe = np.mean(close_genes_fe,axis=0)\n",
    "    \n",
    "        euc_dist_fe_close = distance.euclidean(fe_target_gene,m_close_genes_fe)\n",
    "        cart_d_fe_close = sum(fe_target_gene*m_close_genes_fe)\n",
    "        corr_p_fe_close = pearsonr(fe_target_gene,m_close_genes_fe)[0]\n",
    "        corr_s_fe_close = stats.spearmanr(fe_target_gene,m_close_genes_fe)[0]\n",
    "        L1_dist_fe_close = sum(np.abs(fe_target_gene-m_close_genes_fe))\n",
    "        ks_dist_fe_close = ks_2samp(fe_target_gene,m_close_genes_fe)[0]\n",
    "         \n",
    "    return(euc_dist_fe_essen,corr_p_fe_essen,corr_s_fe_essen,ks_dist_fe_essen,cart_d_fe_essen,L1_dist_fe_essen,euc_dist_fe_close,corr_p_fe_close,corr_s_fe_close,ks_dist_fe_close,cart_d_fe_close,L1_dist_fe_close)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
