{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Update 07.08.20: The features that are related to the target gene will be calculated seperately.** Thus, the following code is NOT using the target sequences that apear here. \n",
    "\n",
    "**The notebook that includes this calculation is \"Target Gene Features.ipnb\"**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## import libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Data.CodonTable import unambiguous_dna_by_id\n",
    "from Bio.Data import CodonTable\n",
    "from itertools import chain\n",
    "import numpy as np\n",
    "from Bio.SeqUtils.CodonUsage import CodonsDict\n",
    "from Bio.Data import CodonTable\n",
    "from collections import Counter\n",
    "import re\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from os import chdir\n",
    "from statistics import mean, pstdev\n",
    "from scipy.stats.mstats import gmean\n",
    "from CAI import CAI, relative_adaptiveness\n",
    "from operator import itemgetter, attrgetter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CAI calculation references:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* the following package, written by Benjamin Lee, is availble online: https://github.com/Benjamin-Lee/CodonAdaptationIndex\n",
    "it is an implementation of Sharp and Li's 1987 formulation of the codon adaption index.\n",
    "* the source code (reading is recommend!): https://github.com/Benjamin-Lee/CodonAdaptationIndex/blob/master/CAI/CAI.py\n",
    "* reference to the paper written by Benjamin Lee, discussing the implementation of this package: https://joss.theoj.org/papers/10.21105/joss.00905\n",
    "* documentation of this package: https://buildmedia.readthedocs.org/media/pdf/cai/latest/cai.pdf\n",
    "\n",
    "attention: the latest version is 1.0.3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install CAI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Methods for Degenerated Codons (ENC calculation)\n",
    "#### http://peterjc.github.io/wiki/Degenerated_Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(CodonTable.standard_dna_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def altcodons(codon, table):\n",
    "    \"\"\"List codons that code for the same aminonacid / are also stop. \n",
    "    Given a codon, the output is a list of the alternative (synonymous) codons.\n",
    "\n",
    "    @param codon\n",
    "    @table code table id (1 is standard)\n",
    "    @return list of codons\n",
    "\n",
    "    \"\"\"\n",
    "    tab = unambiguous_dna_by_id[table]\n",
    "\n",
    "    if codon in tab.stop_codons:\n",
    "        return tab.stop_codons\n",
    "\n",
    "    try:\n",
    "        aa = tab.forward_table[codon]\n",
    "    except:\n",
    "        return []\n",
    "\n",
    "    return [k for (k, v) in tab.forward_table.items()\n",
    "            if v == aa]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def degeneration(codon, table):\n",
    "    \"\"\"Determine how many codons code for the same amino acid / are also stop (degeneration degree of the input codon)\n",
    "\n",
    "    @param codon the codon\n",
    "    @param table code table id\n",
    "    @return the number of codons also coding for the amino acid codon codes for\n",
    "\n",
    "    \"\"\"\n",
    "    return len(altcodons(codon, table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_table():\n",
    "    \"\"\"Define the table that maps each amino acid to its synonymus codons (without stop codons).\n",
    "    @return a dictionary, where the keys are amino acids and the values are the codons.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    standard_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n",
    "    reverse_genetic_code = standard_table.back_table.copy()\n",
    "    # aa = reverse_genetic_code.keys()\n",
    "    # codons = reverse_genetic_code.values()\n",
    "\n",
    "    for aa in reverse_genetic_code.keys():\n",
    "        curr_amino_acid = aa\n",
    "        primary_codon = reverse_genetic_code[aa]\n",
    "        syno_codons = altcodons(primary_codon, 1)\n",
    "        reverse_genetic_code[curr_amino_acid] = syno_codons\n",
    "\n",
    "    reverse_genetic_code.pop(None, None)\n",
    "    return reverse_genetic_code\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'K': ['AAA', 'AAG'], 'N': ['AAT', 'AAC'], 'T': ['ACT', 'ACC', 'ACA', 'ACG'], 'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'], 'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'], 'I': ['ATT', 'ATC', 'ATA'], 'M': ['ATG'], 'Q': ['CAA', 'CAG'], 'H': ['CAT', 'CAC'], 'P': ['CCT', 'CCC', 'CCA', 'CCG'], 'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'E': ['GAA', 'GAG'], 'D': ['GAT', 'GAC'], 'A': ['GCT', 'GCC', 'GCA', 'GCG'], 'G': ['GGT', 'GGC', 'GGA', 'GGG'], 'V': ['GTT', 'GTC', 'GTA', 'GTG'], 'Y': ['TAT', 'TAC'], 'C': ['TGT', 'TGC'], 'W': ['TGG'], 'F': ['TTT', 'TTC']}\n"
     ]
    }
   ],
   "source": [
    "tab = define_table()\n",
    "print(tab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_codons(gene_seq):\n",
    "    \"\"\"For each codon in gene_seq, the function calculates the number of apperances of the codon. \n",
    "    The stop codons are not counted for because we don't use them for ENC calculation.\n",
    "    \n",
    "    @param gene_seq the sequence of the gene (not string type!)\n",
    "    @return a dictionary where the keys are codons and the values are the number of appereances for the codon in the seq.\n",
    "    if the input is illegal sequence (stop codon in the middle of the orf) than the output is empty dictionary.\n",
    "    \n",
    "    \"\"\"\n",
    "    #counts = collections.defaultdict(int)\n",
    "    error_dict = CodonsDict.copy() #empty dictionary\n",
    "    error_dict = {x:np.NaN for x in error_dict}\n",
    "    counts = CodonsDict.copy()\n",
    "    [counts.pop(k) for k in altcodons(\"TAA\",1)] #remove stop codons\n",
    "    [error_dict.pop(k) for k in altcodons(\"TAA\",1)] #remove stop codons\n",
    "    for i in range(0,len(gene_seq),3):\n",
    "        codon = gene_seq[i:i+3]\n",
    "        #if there is a stop codon in the middle of the sequence, return empty dictionary, else continue count:\n",
    "        if codon in altcodons(\"TAA\",1): return error_dict #illegal sequence\n",
    "        else: counts[codon] += 1\n",
    "    return counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_relative_freq(gene_seq):\n",
    "    \"\"\"For each codon, the function calculates its relative frequency,\n",
    "    i.e the number of apperances of the codon,\n",
    "    divided by the number of appereances of the amino-acid that the codon encodes.\n",
    "    if the codon is not in the sequence, its relative frequency is set to NaN.\n",
    "    \n",
    "    @param gene_seq the sequence of the gene (not string type!)\n",
    "    @return a dictionary where the keys are codons and the values are the relative frquencies.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    counts = count_codons(gene_seq)\n",
    "    gene_seq\n",
    "    genetic_code = unambiguous_dna_by_id[1].forward_table #keys = codons; values = Amino acids\n",
    "    reverse_genetic_code = define_table() #keys = Amino acids, values = codons\n",
    "    protein_seq = gene_seq.translate() \n",
    "    return {codon: float(counts[codon])/protein_seq.count(genetic_code[codon]) if protein_seq.count(genetic_code[codon])!=0 else np.NaN \n",
    "            for codon in counts.keys()}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### functions that calculate the demand of codons in the transcriptom (for NTE calculation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_demand_per_gene(counts,gene_id,gene_seq,gene_mrna):\n",
    "    \"\"\"The codon demand for gene 'g' is defined by Ng*Mg: \n",
    "    * the number of apperances of the codon (Ng, unique for the codon and the gene). \n",
    "    * mRNA abundance of the gene (Mg, unique for the gene).\n",
    "    for each codon in gene_seq, the function calculates the demand per gene (Ng*Mg) and updates it to the counts dictionary\n",
    "    \n",
    "    @param counts a dictionary where the keys are the codons and the values are the demands\n",
    "    @param gene_id the gene ORF name\n",
    "    @param gene_seq the sequence of the gene (seq type, not string type)\n",
    "    @param gene_mrna the mRNA abundance of the gene\n",
    "    @return the updated 'counts' dictionary.\n",
    "    \"\"\"\n",
    "    #counts = CodonsDict.copy() #ng*mg \n",
    "    for i in range(0,len(gene_seq),3):\n",
    "        codon = gene_seq[i:i+3]\n",
    "        try:\n",
    "            counts[codon] += gene_mrna\n",
    "            #counts[codon] += mRNA_data[gene_id]\n",
    "        except:\n",
    "            counts[codon] += 0\n",
    "    return counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_demand(sequences):\n",
    "    \"\"\"Total codon demand is defined by how many times it appears in the transcriptom (all the genes of the organism)\n",
    "    The formula for codon i: demand_{i} = \\sum_{g}(Ng_i*Mg)\n",
    "    \"\"\"\n",
    "    u_counts = CodonsDict.copy() #ng*mg \n",
    "    for seq in sequences:\n",
    "        u_counts = codon_demand_per_gene(u_counts,seq.id,seq.seq,seq.mrna)\n",
    "    return u_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## functions that calculate the CUB measures:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ENC_calc(gene_seq):\n",
    "    \"\"\"The effective number of codons (ENC; Wright, 1990) measures the deviation from equal use of synonymous codons.\n",
    "    The ENC is the number of codons that, if used equally, would generate the level of codon bias observed.\n",
    "    ENC takes values from 20.0 (maximum bias) to 61.0 (no bias).\n",
    "    Article: https://eclass.uoa.gr/modules/document/file.php/D473/%CE%92%CE%B9%CE%B2%CE%BB%CE%B9%CE%BF%CE%B3%CF%81%CE%B1%CF%86%CE%AF%CE%B1/DNA%20Composition/Wright_1990.pdf\n",
    "    \n",
    "    The formulas:\n",
    "    ENC =  2+ \\frac{9}{F2} + \\frac{1}{F3} + \\frac{5}{F4} + \\frac{3}{F6} \n",
    "    f_k = \\sum_j^k(H_j) for k=2,3,4,6 (degeneration degree)\n",
    "    H_j = \\frac{n*\\sum_i^k(p_i^2)-1}{n-1} for i=1,...k (synonymus codon of amino acid aa)\n",
    "    \n",
    "    Fk is the average homozygosity (Hj) of amino acids 'j' with degeneration degree k.\n",
    "    Hj is the homozygosity of amino acid 'j' and is calculated from the square codon frequencies. \n",
    "    \n",
    "    @param gene_seq the sequence (not a string type!)\n",
    "    @return ENC index\n",
    "    \"\"\"\n",
    "    \n",
    "    gene_seq = gene_seq[0:-3] #remove stop codon:\n",
    "    freqs = dict() #sum of frequencies for each AA\n",
    "    dd = dict() # this dictionary will group the AA with the same degeneration degrees. the values will be the homozygosities\n",
    "    dd[\"1\"] = []\n",
    "    dd[\"2\"] = []\n",
    "    dd[\"3\"] = []\n",
    "    dd[\"4\"] = []\n",
    "    dd[\"6\"] = []\n",
    "    relative_freq = codon_relative_freq(gene_seq) #a dictionary of relative frequencies for each codon\n",
    "    codon_counter = count_codons(gene_seq) #a dictionary of number of appereances for each codon\n",
    "    reverse_genetic_code = define_table() # define the table that maps each amino acid to its synonymus codons (withoue STOP)\n",
    "    for aa in reverse_genetic_code.keys():\n",
    "        sum_p_power = 0\n",
    "        pow_sum = 0\n",
    "        n = 0\n",
    "        synonyms_codons = reverse_genetic_code[aa] \n",
    "        for codon in synonyms_codons:\n",
    "            sum_p_power += pow(relative_freq[codon],2)\n",
    "            n += codon_counter[codon]\n",
    "            deg_aa = str(degeneration(codon,1))\n",
    "        try:\n",
    "            effective_number_for_aa = ((n*sum_p_power)-1)/(n-1) #Homozygosity of AA (Hj)\n",
    "        except ZeroDivisionError:\n",
    "            #Rarely used aa are those aa for which either the numerator or denominator of \"effective_number_for_aa\" is 0.\n",
    "            #if this occurs they should be treated as absent (place NaN)\n",
    "            effective_number_for_aa = np.NaN\n",
    "        freqs[aa] = effective_number_for_aa\n",
    "        dd[deg_aa].append(effective_number_for_aa);\n",
    "\n",
    "    # When there are no amino acids in a synonymous family, Nc is not calculated as the gene is either too short\n",
    "    # or has extremely skewed amino acid usage.\n",
    "    F2 = np.nanmean(dd[\"2\"])\n",
    "    F3 = np.nanmean(dd[\"3\"])\n",
    "    F4 = np.nanmean(dd[\"4\"])\n",
    "    F6 = np.nanmean(dd[\"6\"])\n",
    "\n",
    "    #if isoleucine (I) is missing, then F3 should be calculated as the mean of F2 and F4.\n",
    "    if np.isnan(F3): \n",
    "        F3 = (F2+F4)/2\n",
    "\n",
    "    #If any of the other S-F types are completely missing or 'rarely used' then the gene is probably too short\n",
    "    #(e.g., less than 61 codons) to accurately measure SCU bias, or exhibits extremely skewed aa usage. \n",
    "    if (F2==0) | (F3==0) | (F4==0) | (F6==0): \n",
    "        ENC = np.NaN\n",
    "    else:\n",
    "        ENC = 2+ (9/F2) + (1/F3) + (5/F4) + (3/F6)\n",
    "    \n",
    "    #The value of ENC can be greater than 61 if the observed codon usage pattern is more uniform than expected by chance.\n",
    "    #Such a rare event is most likely when aa composition is very extreme (e.g., such that the Fs are obtained\n",
    "    #by averaging over very few aa), and the gene is very short. In these cases, the value of ENC should be revised to 61. \n",
    "    if ENC>61:\n",
    "        ENC=61\n",
    "    \n",
    "    return ENC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tAI_profile(sequence, weights):\n",
    "    \"\"\"creates a list of the imported weights for the sequence, counting all codons.\"\"\" \n",
    "    sequence = sequence.upper()\n",
    "    sequence = [sequence[i : i + 3] for i in range(0, len(sequence), 3)]\n",
    "    sequence_weights = []\n",
    "    for codon in sequence:\n",
    "        sequence_weights.append(weights[codon])\n",
    "    return sequence_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_NTE_weights(gene_data,mRNA_data,tAI_weights):\n",
    "    \"\"\"NTE is normalized translation efficiency. it it defined as the supply/demand ratio,\n",
    "    which is why the weight of each codon is calculated as tAI/demand.\"\"\"\n",
    "    for i in range(len(gene_data)):\n",
    "        try:\n",
    "            gene_data[i].mrna = mRNA_data[gene_data[i].id]\n",
    "        except:\n",
    "            gene_data[i].mrna = None\n",
    "    cu = codon_demand(gene_data) #demand of each codon\n",
    "    cu_norm = {k: v/max(cu.values()) for k, v in cu.items()} #normalized demand of each codon\n",
    "    nte = {k: v/cu_norm[k] for k, v in tAI_weights.items()} #the supply/demand ratio.\n",
    "    # we assume that the supply is correlative to the tAI becaus we dont have the tRNA levels of each codon in the cell.\n",
    "    NTE_weights = {k: v/max(nte.values()) for k, v in nte.items()} #normalized NTE for each codon\n",
    "    return NTE_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def NTE_profile(sequence, weights):\n",
    "    \"\"\"create a list of the calculated weights for the sequence, counting all codons.\"\"\"\n",
    "    sequence = sequence.upper()\n",
    "    sequence = [sequence[i : i + 3] for i in range(0, len(sequence), 3)] #cut the sequence to codons\n",
    "    sequence_weights = []\n",
    "    for codon in sequence:\n",
    "        sequence_weights.append(weights[codon])\n",
    "    return sequence_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_RCBS_per_gene(gene_seq):\n",
    "    \"\"\"The function calculates the RCBS of the gene, which is defined as: RCBS = geomean([d_xyz+1])-1\n",
    "    For each codon we calculate the RCBS weight, which is defined as d_xyz = [f(x,y,z)-f1(x)f2(y)f3(z)]/[f1(x)f2(y)f3(z)]\n",
    "    where f(x,y,z) is the normalized frequency of the codon xyz (number of apperances divided with length of the sequence),\n",
    "    and f1(x) is the normalized frequency of the nucleotide x in the first position of the codons in the seq.\n",
    "    From these weights (which are unique for the gene), we create a profile of d_xyz+1, calculate the geometric mean,\n",
    "    and substract 1. This leads to the RCBS index of the sequence.\n",
    "    Original article: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2646356/\"\"\"\n",
    "\n",
    "    #counts = CodonsDict.copy() #ng*mg\n",
    "    codons_list = (gene_seq[i:i + 3].upper() for i in range(0, len(gene_seq), 3))\n",
    "    codons_list = list(codons_list) #a flat list of all the codons in the sequence. the length of the list is len(seq)/3\n",
    "    counts = Counter(codons_list) #dictionary sub-class\n",
    "\n",
    "    L = len(gene_seq)/3\n",
    "    Fxyz = {k: v/L for k, v in counts.items()} #the normalized frquency of the codon\n",
    "    \n",
    "    F1 = dict.fromkeys(['A','C','G','T']) #here we insert the observed freq of each nuc in the first codon position\n",
    "    F1['A'] = len(list((val for val in codons_list if re.findall(\"\\AA\", val))))/L\n",
    "    F1['C'] = len(list((val for val in codons_list if re.findall(\"\\AC\", val))))/L\n",
    "    F1['G'] = len(list((val for val in codons_list if re.findall(\"\\AG\", val))))/L\n",
    "    F1['T'] = len(list((val for val in codons_list if re.findall(\"\\AT\", val))))/L\n",
    "    \n",
    "    F2 = dict.fromkeys(['A','C','G','T']) #here we insert the observed freq of each nuc in the second codon position\n",
    "    F2['A'] = len(list((val for val in codons_list if val[1]=='A')))/L\n",
    "    F2['C'] = len(list((val for val in codons_list if val[1]=='C')))/L\n",
    "    F2['G'] = len(list((val for val in codons_list if val[1]=='G')))/L\n",
    "    F2['T'] = len(list((val for val in codons_list if val[1]=='T')))/L\n",
    "    \n",
    "    F3 = dict.fromkeys(['A','C','G','T']) #here we insert the observed freq of each nuc in the third codon position\n",
    "    F3['A'] = len(list((val for val in codons_list if re.findall(\"A$\", val))))/L\n",
    "    F3['C'] = len(list((val for val in codons_list if re.findall(\"C$\", val))))/L\n",
    "    F3['G'] = len(list((val for val in codons_list if re.findall(\"G$\", val))))/L\n",
    "    F3['T'] = len(list((val for val in codons_list if re.findall(\"T$\", val))))/L\n",
    "    \n",
    "    ### calculaute the RCBS weights per codon (d_xyz):\n",
    "    RCBS_weights = CodonsDict.copy()\n",
    "    for codon in RCBS_weights.keys():\n",
    "        x = codon[0]\n",
    "        y = codon[1]\n",
    "        z = codon[2]\n",
    "        try:\n",
    "            RCBS_weights[codon] = (Fxyz[codon]-F1[x]*F2[y]*F3[z])/(F1[x]*F2[y]*F3[z]) #d_xyz definition\n",
    "        except KeyError:\n",
    "            RCBS_weights[codon] = 0 #the codon does not apprear in the sequence, hence its weight is set to zero.\n",
    "    \n",
    "    ## calculate the profile, for RCBS calculation: \n",
    "    sequence_profile = [] \n",
    "    for codon in codons_list:\n",
    "        sequence_profile.append(1+RCBS_weights[codon])\n",
    "    \n",
    "    #Now sequence_weights is a list of (1+d_xyz), so we can calculate the geoman-1, thus achive RCBS index.\n",
    "    RCBS_gene = gmean(sequence_profile)-1\n",
    "    return RCBS_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_RCA_weights(ref_seq):\n",
    "    \"\"\"The function calculates the RCA weights per codon in the reference genome. The weight of codon xyz is defined as:\n",
    "    RCA_xyz = [f(x,y,z)]/[f1(x)f2(y)f3(z)]\n",
    "    where f(x,y,z) is the normalized frequency of the codon xyz (number of apperances divided with length of the sequences),\n",
    "    and f1(x) is the normalized frequency of the nucleotide x in the first position of the codons in the seq.\n",
    "    From these weights (which are calculated based on whole genome), we create a profile of RCA_xyz (see function \n",
    "    RCA_profile). This leads to the RCA index of the sequence.\n",
    "    Original article: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2885275/\"\"\"\n",
    "    \n",
    "    ## count the number of apperances of each codon in the reference set of genes:\n",
    "    sequences = (list((str(sequence[i:i + 3].upper())) for i in range(0, len(sequence), 3)) for sequence in ref_seq)\n",
    "    codons_list = list(sequences) # flat list of all codons (to be used for counting)\n",
    "    codons_list = list(chain(*codons_list)) # flat list of all codons (to be used for counting)\n",
    "    counts = Counter(codons_list)\n",
    "\n",
    "    L = len(codons_list)\n",
    "    CHIxyz = {str(k): v/L for k, v in counts.items()} #the normalized frquency of the codon\n",
    "    \n",
    "    CHI1 = dict.fromkeys(['A','C','G','T']) #here we insert the observed freq of each nuc in the first codon position\n",
    "    CHI1['A'] = len(list((val for val in codons_list if re.findall(\"\\AA\", val))))/L\n",
    "    CHI1['C'] = len(list((val for val in codons_list if re.findall(\"\\AC\", val))))/L\n",
    "    CHI1['G'] = len(list((val for val in codons_list if re.findall(\"\\AG\", val))))/L\n",
    "    CHI1['T'] = len(list((val for val in codons_list if re.findall(\"\\AT\", val))))/L\n",
    "    \n",
    "    CHI2 = dict.fromkeys(['A','C','G','T']) #here we insert the observed freq of each nuc in the second codon position\n",
    "    CHI2['A'] = len(list((val for val in codons_list if val[1]=='A')))/L\n",
    "    CHI2['C'] = len(list((val for val in codons_list if val[1]=='C')))/L\n",
    "    CHI2['G'] = len(list((val for val in codons_list if val[1]=='G')))/L\n",
    "    CHI2['T'] = len(list((val for val in codons_list if val[1]=='T')))/L\n",
    "    \n",
    "    CHI3 = dict.fromkeys(['A','C','G','T']) #here we insert the observed freq of each nuc in the third codon position\n",
    "    CHI3['A'] = len(list((val for val in codons_list if re.findall(\"A$\", val))))/L\n",
    "    CHI3['C'] = len(list((val for val in codons_list if re.findall(\"C$\", val))))/L\n",
    "    CHI3['G'] = len(list((val for val in codons_list if re.findall(\"G$\", val))))/L\n",
    "    CHI3['T'] = len(list((val for val in codons_list if re.findall(\"T$\", val))))/L\n",
    "    \n",
    "    ### calculaute the RCBS weights per codon (d_xyz):\n",
    "    RCA_weights = CodonsDict.copy()\n",
    "    for codon in RCA_weights.keys():\n",
    "        x = codon[0]\n",
    "        y = codon[1]\n",
    "        z = codon[2]\n",
    "        try:\n",
    "            RCA_weights[codon] = (CHIxyz[codon])/(CHI1[x]*CHI2[y]*CHI3[z]) #d_xyz definition\n",
    "        except KeyError:\n",
    "            RCA_weights[codon] = 0 #the codon does not apprear in the reference sequence, hence its weight is set to zero.\n",
    "    return RCA_weights\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def RCA_profile(sequence,weights):\n",
    "    \"\"\"create a list of the calculated weights for the sequence, counting all codons.\"\"\"\n",
    "    sequence = sequence.upper()\n",
    "    sequence = [sequence[i : i + 3] for i in range(0, len(sequence), 3)] #cut the sequence to codons\n",
    "    sequence_weights = []\n",
    "    for codon in sequence:\n",
    "        sequence_weights.append(weights[codon])\n",
    "    return sequence_weights"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data loading and CUB calculation:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## data loading:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = r'/home/karinsio/my_project_dir/iGEM/Genes_Features_seq'\n",
    "chdir(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'K': 'AAG', 'N': 'AAT', 'T': 'ACT', 'R': 'CGT', 'S': 'TCT', 'I': 'ATT', 'M': 'ATG', 'Q': 'CAG', 'H': 'CAT', 'P': 'CCT', 'L': 'TTG', 'E': 'GAG', 'D': 'GAT', 'A': 'GCT', 'G': 'GGT', 'V': 'GTT', 'Y': 'TAT', 'C': 'TGT', 'W': 'TGG', 'F': 'TTT', None: 'TAA'}\n"
     ]
    }
   ],
   "source": [
    "standard_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n",
    "print(standard_table.back_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_orf_coding():\n",
    "### get sequence data\n",
    "    with gzip.open(genetic_encoding_file, 'rt') as data:\n",
    "        gene_data_list = [seq for seq in SeqIO.parse(data, 'fasta') if len(seq)% 3 == 0]\n",
    "    return gene_data_list;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_refernce_seq_wholeGenome():\n",
    "    \"\"\"get sequences of all the yeast genome, to use as refercnce for calculation of CAI and RCA\"\"\"\n",
    "    with gzip.open(genetic_encoding_file, 'rt') as data:\n",
    "        #ref is only the sequence of the genes that are divided by 3\n",
    "        ref = [seq.seq for seq in SeqIO.parse(data, 'fasta') if len(seq)% 3 == 0]\n",
    "    return ref;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_PA_data():\n",
    "    \"\"\"read the Protein Abundance (PA) file and extract the PA of each protein (dictionary)\"\"\"\n",
    "    \n",
    "    PA_filename = '4932-GPM_2012_09_Saccharomyces_cerevisiae.txt'\n",
    "    PA_data = pd.read_csv(PA_filename,header=11,delimiter='\\t') #DataFrame\n",
    "    PA_data = PA_data.drop(columns=['#internal_id','raw_spectral_count']).rename(columns={\"string_external_id\": \"ORF\"})\n",
    "    PA_data=PA_data.set_index('ORF').to_dict()\n",
    "    PA_data=PA_data['abundance'] #dictionay, the keys are ORF names and the values are abundance level of protein.\n",
    "    PA_data = {k[5:]: v for k, v in PA_data.items()} #cut the \"4932.\" chars from the protein ORF name.\n",
    "    return PA_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "PA_data = import_PA_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_reference_seq_HEG(PA_data):\n",
    "    \"\"\"import the highly expressed genes sequences (2% of the genome with the higest protein abundance)\n",
    "    to use as referance seq for calculation of CAI and RCA.\"\"\"\n",
    "    \n",
    "    sorted_PA = sorted(list(PA_data.items()), key=itemgetter(1), reverse=True) #sorted list from high to low PA\n",
    "    highest_PA = sorted_PA[0:round(len(sorted_PA)*0.02)] #list of tuples (ORF,PA) containing 20% of the genes with highly protain abundance \n",
    "    HEG_id = list(map(itemgetter(0), highest_PA)) #list of the ORF of the highly expressed genes (HEG)\n",
    "    ### get the sequences of the highly expressed genes (HEG, account for 2% of the genome, have highest PA),\n",
    "    ### to use as reference for calculation of CAI:\n",
    "    with gzip.open(genetic_encoding_file, 'rt') as data:\n",
    "        #ref is only the sequence of the genes that are included in the list of HEG_id\n",
    "        ref_HEG = [seq.seq for seq in SeqIO.parse(data, 'fasta') if HEG_id.count(seq.id) == 1]\n",
    "    return ref_HEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "## coding sequence for each gene in yeast:\n",
    "genetic_encoding_file = r'orf_coding_all.fasta.gz'\n",
    "gene_data = import_orf_coding()\n",
    "\n",
    "## referance sequences: \n",
    "ref_seq_wholeGenome = import_refernce_seq_wholeGenome()\n",
    "ref_seq_HEG = import_reference_seq_HEG(PA_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_tAI_weights():\n",
    "    \"\"\"read the tAI file and extract the weight for each codon (dictionary)\"\"\"\n",
    "    \n",
    "    tAI_file = r'1-s2.0-S0092867410003193-mmc2.xls'\n",
    "    xls = pd.ExcelFile(tAI_file)\n",
    "    df1 = pd.read_excel(xls, 'tAI',header=5,skiprows=(70,71))\n",
    "    tAI_df = df1.drop(columns=['anti-codon','Human',' E. coli K12','C. elegance','D. melanogaster','Aeropyrum pernix ']).rename(columns={\"S. Cerevisiae\": \"tAI_weight\"}).set_index('Codon')\n",
    "    tAI_weights = tAI_df.to_dict()\n",
    "    tAI_weights=tAI_weights['tAI_weight'] #dictionary, the keys are codons and the values are weights for tAI calculation.\n",
    "    return tAI_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_mrna_data():\n",
    "    \"\"\"read the mRNA file and extract the demand of each codon, and then the NTE for each codon (dictionary)\"\"\"\n",
    "    \n",
    "    mRNA_filename = 'GSE75897_RiboZero_RPKMs.txt'\n",
    "    mRNA_data = pd.read_csv(mRNA_filename,header=None,delimiter='\\t')\n",
    "    mRNA_data=mRNA_data.set_index(0).to_dict()\n",
    "    mRNA_data=mRNA_data[1] #dictionay, the keys are ORF names and the values are mRNA levels (in RPKM)\n",
    "    return mRNA_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNA_data = import_mrna_data()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cub calculation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CUB_calc(gene_data_list,ref_seq_wholeGenome,ref_seq_HEG,mRNA_data,PA_data):\n",
    "    \"\"\"The Main Function!\n",
    "    @return a DataFrame of the genes and their codon usage bias analysis.\n",
    "    \"\"\"\n",
    "    winSize = 48 #window size in nucleotides\n",
    "    winStep = 1 #window step in nucleotides\n",
    "    numWin = 100 #number of windows in the gene. \n",
    "    # if gene length is smaller than 150 nuc, than the window value will be placed with nan\n",
    "    \n",
    "    # calculate the weights for CAI: the reference genome is all the yeast genome\n",
    "    CAI_weights_wholeGenome = relative_adaptiveness(sequences=ref_seq_wholeGenome)\n",
    "    # calculate the weights for CAI: the reference genome is the highly expressed genes (HEG) with the highest PA: \n",
    "    CAI_weights_HEG = relative_adaptiveness(sequences=ref_seq_HEG)\n",
    "    #calculate the tAI weights:\n",
    "    tAI_weights = import_tAI_weights()\n",
    "    # calculate the NTE weights (based on the whole organism):\n",
    "    NTE_weights = calc_NTE_weights(gene_data_list,mRNA_data,tAI_weights)\n",
    "    #calculate the weights for RCA: the reference genome is all the yeast genome\n",
    "    RCA_weights_wholeGenome = calc_RCA_weights(ref_seq=ref_seq_wholeGenome)\n",
    "    #calculate the weights for RCA: the reference genome is the highly expressed genes (HEG) with the highest PA: \n",
    "    RCA_weights_HEG = calc_RCA_weights(ref_seq=ref_seq_HEG)    \n",
    "    \n",
    "    all_genes = dict()\n",
    "    for gene in gene_data_list:\n",
    "        curr_gene = dict()\n",
    "         ### gene sequence\n",
    "        curr_gene['gene_sequence'] = str(gene.seq)\n",
    "        ### length of nuc sequence\n",
    "        curr_gene['length'] = len(str(gene.seq))\n",
    "        ### gene ID \n",
    "        curr_gene['ORF'] = str(gene.id)\n",
    "        ### gene mRNA levels:\n",
    "        try:\n",
    "            curr_gene['mRNA_RPKM'] = mRNA_data[str(gene.id)]\n",
    "        except:\n",
    "            curr_gene['mRNA_RPKM'] = np.NaN\n",
    "        ### gene Protein abundance (PA) levels:\n",
    "        try:\n",
    "            curr_gene['PA'] = PA_data[str(gene.id)]\n",
    "        except:\n",
    "            curr_gene['PA'] = np.NaN\n",
    "                \n",
    "        ### gene ENC:\n",
    "        curr_gene['ENC'] = ENC_calc(gene.seq)\n",
    "        ### gene CAI, ref=wholeGenome: \n",
    "        cai_wholeGenome = CAI(str(gene.seq), weights=CAI_weights_wholeGenome)\n",
    "        curr_gene['CAI_index_wholeGenome'] = cai_wholeGenome[0];\n",
    "        #curr_gene['CAI_profile_wholeGenome'] = cai_wholeGenome[1];\n",
    "        curr_gene['CAI_STD_wholeGenome'] = float(pstdev(cai_wholeGenome[1])); #standard deviation\n",
    "        curr_gene['CAI_first_50_codons_wholeGenome'] = float(gmean(cai_wholeGenome[1][0:50]));\n",
    "        curr_gene['CAI_last_50_codons_wholeGenome'] = float(gmean(cai_wholeGenome[1][-50:])); \n",
    "        ### gene CAI, ref=HEG: \n",
    "        cai_HEG = CAI(str(gene.seq), weights=CAI_weights_HEG)\n",
    "        curr_gene['CAI_index_HEG'] = cai_HEG[0];\n",
    "        #curr_gene['CAI_profile_HEG'] = cai_HEG[1];\n",
    "        curr_gene['CAI_STD_HEG'] = float(pstdev(cai_HEG[1]));\n",
    "        curr_gene['CAI_first_50_codons_HEG'] = float(gmean(cai_HEG[1][0:50]));\n",
    "        curr_gene['CAI_last_50_codons_HEG'] = float(gmean(cai_HEG[1][-50:])); \n",
    "        ### gene tAI:\n",
    "        tai_output = tAI_profile(str(gene.seq), weights=tAI_weights)\n",
    "        curr_gene['tAI_index'] = float(gmean(tai_output))\n",
    "        #curr_gene['tAI_profile'] = tai_output\n",
    "        curr_gene['tAI_STD'] = float(pstdev(tai_output))\n",
    "        curr_gene['tAI_first_50_codons'] = float(gmean(tai_output[0:50]));\n",
    "        curr_gene['tAI_last_50_codons'] = float(gmean(tai_output[-50:]));        \n",
    "        ### gene NTE:\n",
    "        nte_output = NTE_profile(str(gene.seq), weights=NTE_weights)\n",
    "        curr_gene['NTE_index'] = float(gmean(nte_output))\n",
    "        #curr_gene['NTE_profile'] = nte_output\n",
    "        curr_gene['NTE_STD'] = float(pstdev(nte_output))\n",
    "        curr_gene['NTE_first_50_codons'] = float(gmean(nte_output[0:50]));\n",
    "        curr_gene['NTE_last_50_codons'] = float(gmean(nte_output[-50:]));   \n",
    "\n",
    "        ### gene RCBS: \n",
    "        curr_gene['RCBS'] = calc_RCBS_per_gene(str(gene.seq))\n",
    "        ### gene RCA, ref = wholeGenome:\n",
    "        rca_wholeGenome = RCA_profile(str(gene.seq),weights=RCA_weights_wholeGenome)\n",
    "        curr_gene['RCA_index_wholeGenome'] = float(gmean(rca_wholeGenome))\n",
    "        #curr_gene['RCA_profile_wholeGenome'] = rca_wholeGenome\n",
    "        curr_gene['RCA_STD_wholeGenome'] = float(pstdev(rca_wholeGenome))\n",
    "        curr_gene['RCA_first_50_codons_wholeGenome'] = float(gmean(rca_wholeGenome[0:50]));\n",
    "        curr_gene['RCA_last_50_codons_wholeGenome'] = float(gmean(rca_wholeGenome[-50:]));   \n",
    "        ### gene RCA, ref = wholeGenome:\n",
    "        rca_HEG = RCA_profile(str(gene.seq),weights=RCA_weights_HEG)\n",
    "        curr_gene['RCA_index_HEG'] = float(gmean(rca_HEG))\n",
    "        #curr_gene['RCA_profile_HEG'] = rca_HEG\n",
    "        curr_gene['RCA_STD_HEG'] = float(pstdev(rca_HEG))\n",
    "        curr_gene['RCA_first_50_codons_HEG'] = float(gmean(rca_HEG[0:50]));\n",
    "        curr_gene['RCA_last_50_codons_HEG'] = float(gmean(rca_HEG[-50:]));\n",
    "        \n",
    "        ########## windowing:\n",
    "        gene_seq = gene.seq\n",
    "        for i in range(1,numWin+1):\n",
    "            win = gene_seq[i:i+winSize] #seq\n",
    "            if len(win)==winSize:\n",
    "                ### window CAI, ref=wholeGenome: \n",
    "                caiWin_wholeGenome = CAI(str(win), weights=CAI_weights_wholeGenome)\n",
    "                curr_gene['CAI_index_win'+str(i)+'_wholeGenome'] = caiWin_wholeGenome[0];\n",
    "                ### window CAI, ref=HEG: \n",
    "                caiWin_HEG = CAI(str(win), weights=CAI_weights_HEG)\n",
    "                curr_gene['CAI_index_win'+str(i)+'_HEG'] = caiWin_HEG[0];\n",
    "                ### window tAI:\n",
    "                tai_output_win = tAI_profile(str(win), weights=tAI_weights)\n",
    "                curr_gene['tAI_index_win'+str(i)] = float(gmean(tai_output_win))\n",
    "                ### window NTE:\n",
    "                tai_output_win = NTE_profile(str(win), weights=NTE_weights)\n",
    "                curr_gene['NTE_index_win'+str(i)] = float(gmean(tai_output_win))\n",
    "                ### window RCBS: \n",
    "                curr_gene['RCBS_win'+str(i)] = calc_RCBS_per_gene(str(win))\n",
    "                ### window RCA, ref = wholeGenome:\n",
    "                rcaWin_wholeGenome = RCA_profile(str(win),weights=RCA_weights_wholeGenome)\n",
    "                curr_gene['RCA_index_win'+str(i)+'_wholeGenome'] = float(gmean(rcaWin_wholeGenome))\n",
    "                ### window RCA, ref = wholeGenome:\n",
    "                rcaWin_HEG = RCA_profile(str(win),weights=RCA_weights_HEG)\n",
    "                curr_gene['RCA_index_win'+str(i)+'_HEG'] = float(gmean(rcaWin_HEG))\n",
    "            else:\n",
    "                curr_gene['CAI_index_win'+str(i)+'_wholeGenome']=np.NaN\n",
    "                curr_gene['CAI_index_win'+str(i)+'_HEG']=np.NaN\n",
    "                curr_gene['tAI_index_win'+str(i)]=np.NaN\n",
    "                curr_gene['NTE_index_win'+str(i)]=np.NaN\n",
    "                curr_gene['CAI_index_win'+str(i)+'_wholeGenome']=np.NaN\n",
    "                curr_gene['RCBS_win'+str(i)]=np.NaN\n",
    "                curr_gene['RCA_index_win'+str(i)+'_wholeGenome']=np.NaN\n",
    "                curr_gene['RCA_index_win'+str(i)+'_HEG'] = np.NaN\n",
    "        \n",
    "        ## insert all the features of the current gene into the final table:\n",
    "        all_genes[gene.id] = curr_gene\n",
    "        \n",
    "    \n",
    "    gene_data_df = pd.DataFrame.from_dict(all_genes, orient = 'index').reset_index().rename(columns={'index':'ORF'})\n",
    "    return gene_data_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/karinsio/my_project_dir/my_project_env/lib/python3.6/site-packages/ipykernel_launcher.py:51: RuntimeWarning: Mean of empty slice\n",
      "/home/karinsio/my_project_dir/my_project_env/lib/python3.6/site-packages/ipykernel_launcher.py:50: RuntimeWarning: Mean of empty slice\n",
      "/home/karinsio/my_project_dir/my_project_env/lib/python3.6/site-packages/ipykernel_launcher.py:52: RuntimeWarning: Mean of empty slice\n",
      "/home/karinsio/my_project_dir/my_project_env/lib/python3.6/site-packages/ipykernel_launcher.py:53: RuntimeWarning: Mean of empty slice\n"
     ]
    }
   ],
   "source": [
    "gene_data_df = CUB_calc(gene_data,ref_seq_wholeGenome,ref_seq_HEG,mRNA_data,PA_data)\n",
    "# gene_data_df.to_csv(r'CUB_results_ver2.csv')\n",
    "gene_data_df.to_excel(r'CUB_results_final.xlsx',engine='xlsxwriter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ORF</th>\n",
       "      <th>gene_sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>ORF</th>\n",
       "      <th>mRNA_RPKM</th>\n",
       "      <th>PA</th>\n",
       "      <th>ENC</th>\n",
       "      <th>CAI_index_wholeGenome</th>\n",
       "      <th>CAI_STD_wholeGenome</th>\n",
       "      <th>CAI_first_50_codons_wholeGenome</th>\n",
       "      <th>...</th>\n",
       "      <th>RCBS_win99</th>\n",
       "      <th>RCA_index_win99_wholeGenome</th>\n",
       "      <th>RCA_index_win99_HEG</th>\n",
       "      <th>CAI_index_win100_wholeGenome</th>\n",
       "      <th>CAI_index_win100_HEG</th>\n",
       "      <th>tAI_index_win100</th>\n",
       "      <th>NTE_index_win100</th>\n",
       "      <th>RCBS_win100</th>\n",
       "      <th>RCA_index_win100_wholeGenome</th>\n",
       "      <th>RCA_index_win100_HEG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL001C</td>\n",
       "      <td>ATGGTACTGACGATTTATCCTGACGAACTCGTACAAATAGTGTCTG...</td>\n",
       "      <td>3483</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>37.80</td>\n",
       "      <td>11.60</td>\n",
       "      <td>52.499762</td>\n",
       "      <td>0.734169</td>\n",
       "      <td>0.250486</td>\n",
       "      <td>0.759264</td>\n",
       "      <td>...</td>\n",
       "      <td>2.750944</td>\n",
       "      <td>1.222763</td>\n",
       "      <td>0.787172</td>\n",
       "      <td>0.707072</td>\n",
       "      <td>0.178779</td>\n",
       "      <td>0.241931</td>\n",
       "      <td>0.027772</td>\n",
       "      <td>1.879886</td>\n",
       "      <td>1.032415</td>\n",
       "      <td>0.405449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL002W</td>\n",
       "      <td>ATGGAGCAAAATGGCCTTGACCACGACAGCAGATCTAGCATCGATA...</td>\n",
       "      <td>3825</td>\n",
       "      <td>YAL002W</td>\n",
       "      <td>22.17</td>\n",
       "      <td>7.27</td>\n",
       "      <td>50.928649</td>\n",
       "      <td>0.742818</td>\n",
       "      <td>0.239425</td>\n",
       "      <td>0.714702</td>\n",
       "      <td>...</td>\n",
       "      <td>3.061334</td>\n",
       "      <td>1.092761</td>\n",
       "      <td>0.320760</td>\n",
       "      <td>0.523649</td>\n",
       "      <td>0.093294</td>\n",
       "      <td>0.189421</td>\n",
       "      <td>0.046871</td>\n",
       "      <td>2.131717</td>\n",
       "      <td>0.605164</td>\n",
       "      <td>0.262181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YAL003W</td>\n",
       "      <td>ATGGCATCCACCGATTTCTCCAAGATTGAAACTTTGAAACAATTAA...</td>\n",
       "      <td>621</td>\n",
       "      <td>YAL003W</td>\n",
       "      <td>4621.86</td>\n",
       "      <td>3029.00</td>\n",
       "      <td>26.669428</td>\n",
       "      <td>0.825756</td>\n",
       "      <td>0.190030</td>\n",
       "      <td>0.836821</td>\n",
       "      <td>...</td>\n",
       "      <td>2.118197</td>\n",
       "      <td>1.197576</td>\n",
       "      <td>1.908536</td>\n",
       "      <td>0.640120</td>\n",
       "      <td>0.117762</td>\n",
       "      <td>0.193914</td>\n",
       "      <td>0.025949</td>\n",
       "      <td>2.458412</td>\n",
       "      <td>1.115684</td>\n",
       "      <td>0.371074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YAL004W</td>\n",
       "      <td>ATGGGTGTCACCAGCGGTGGCCTTAACTTCAAAGATACCGTCTTCA...</td>\n",
       "      <td>648</td>\n",
       "      <td>YAL004W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.80</td>\n",
       "      <td>34.435740</td>\n",
       "      <td>0.674925</td>\n",
       "      <td>0.240300</td>\n",
       "      <td>0.659022</td>\n",
       "      <td>...</td>\n",
       "      <td>3.178232</td>\n",
       "      <td>0.991119</td>\n",
       "      <td>0.402804</td>\n",
       "      <td>0.637584</td>\n",
       "      <td>0.123598</td>\n",
       "      <td>0.297283</td>\n",
       "      <td>0.043497</td>\n",
       "      <td>2.118197</td>\n",
       "      <td>0.883690</td>\n",
       "      <td>0.389291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YAL005C</td>\n",
       "      <td>ATGTCAAAAGCTGTCGGTATTGATTTAGGTACAACATACTCGTGTG...</td>\n",
       "      <td>1929</td>\n",
       "      <td>YAL005C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1117.00</td>\n",
       "      <td>28.341978</td>\n",
       "      <td>0.823721</td>\n",
       "      <td>0.195552</td>\n",
       "      <td>0.837872</td>\n",
       "      <td>...</td>\n",
       "      <td>2.391503</td>\n",
       "      <td>1.175160</td>\n",
       "      <td>1.734433</td>\n",
       "      <td>0.665937</td>\n",
       "      <td>0.123181</td>\n",
       "      <td>0.187328</td>\n",
       "      <td>0.027386</td>\n",
       "      <td>2.358922</td>\n",
       "      <td>0.987152</td>\n",
       "      <td>0.399857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>YAL007C</td>\n",
       "      <td>ATGATCAAATCTACAATTGCTCTACCCTCTTTCTTCATTGTTTTAA...</td>\n",
       "      <td>648</td>\n",
       "      <td>YAL007C</td>\n",
       "      <td>212.42</td>\n",
       "      <td>255.00</td>\n",
       "      <td>51.760655</td>\n",
       "      <td>0.777593</td>\n",
       "      <td>0.223742</td>\n",
       "      <td>0.778706</td>\n",
       "      <td>...</td>\n",
       "      <td>2.517394</td>\n",
       "      <td>1.121169</td>\n",
       "      <td>0.682821</td>\n",
       "      <td>0.769939</td>\n",
       "      <td>0.215059</td>\n",
       "      <td>0.304178</td>\n",
       "      <td>0.034379</td>\n",
       "      <td>2.517394</td>\n",
       "      <td>0.990837</td>\n",
       "      <td>0.456100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YAL008W</td>\n",
       "      <td>ATGACTTTGGCTTTTAATATGCAACGGTTGGTGTTTCGTAATTTGA...</td>\n",
       "      <td>597</td>\n",
       "      <td>YAL008W</td>\n",
       "      <td>49.96</td>\n",
       "      <td>72.70</td>\n",
       "      <td>51.661854</td>\n",
       "      <td>0.714422</td>\n",
       "      <td>0.260086</td>\n",
       "      <td>0.661697</td>\n",
       "      <td>...</td>\n",
       "      <td>3.178232</td>\n",
       "      <td>1.069899</td>\n",
       "      <td>0.499223</td>\n",
       "      <td>0.567410</td>\n",
       "      <td>0.081334</td>\n",
       "      <td>0.183980</td>\n",
       "      <td>0.054649</td>\n",
       "      <td>3.382807</td>\n",
       "      <td>0.434412</td>\n",
       "      <td>0.195369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>YAL009W</td>\n",
       "      <td>ATGGAGCCAGAGAGCATAGGCGATGTGGGGAACCATGCCCAGGATG...</td>\n",
       "      <td>780</td>\n",
       "      <td>YAL009W</td>\n",
       "      <td>34.96</td>\n",
       "      <td>37.20</td>\n",
       "      <td>58.167277</td>\n",
       "      <td>0.690552</td>\n",
       "      <td>0.262954</td>\n",
       "      <td>0.579445</td>\n",
       "      <td>...</td>\n",
       "      <td>2.507514</td>\n",
       "      <td>0.969335</td>\n",
       "      <td>0.366193</td>\n",
       "      <td>0.540432</td>\n",
       "      <td>0.087199</td>\n",
       "      <td>0.082978</td>\n",
       "      <td>0.019178</td>\n",
       "      <td>2.507514</td>\n",
       "      <td>0.713440</td>\n",
       "      <td>0.237107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>YAL010C</td>\n",
       "      <td>ATGCTACCCTATATGGACCAAGTACTAAGGGCATTTTATCAGAGCA...</td>\n",
       "      <td>1482</td>\n",
       "      <td>YAL010C</td>\n",
       "      <td>16.67</td>\n",
       "      <td>6.19</td>\n",
       "      <td>50.675034</td>\n",
       "      <td>0.739047</td>\n",
       "      <td>0.240760</td>\n",
       "      <td>0.664406</td>\n",
       "      <td>...</td>\n",
       "      <td>2.236818</td>\n",
       "      <td>1.068921</td>\n",
       "      <td>0.505618</td>\n",
       "      <td>0.803427</td>\n",
       "      <td>0.370478</td>\n",
       "      <td>0.380276</td>\n",
       "      <td>0.044353</td>\n",
       "      <td>1.721829</td>\n",
       "      <td>0.926069</td>\n",
       "      <td>0.619003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>YAL011W</td>\n",
       "      <td>ATGCCTGCTGTCTTGAGAACCAGGTCCAAAGAATCCTCTATAGAGC...</td>\n",
       "      <td>1878</td>\n",
       "      <td>YAL011W</td>\n",
       "      <td>17.40</td>\n",
       "      <td>404.00</td>\n",
       "      <td>53.630576</td>\n",
       "      <td>0.719718</td>\n",
       "      <td>0.246476</td>\n",
       "      <td>0.718603</td>\n",
       "      <td>...</td>\n",
       "      <td>0.721549</td>\n",
       "      <td>1.114492</td>\n",
       "      <td>0.566084</td>\n",
       "      <td>0.498507</td>\n",
       "      <td>0.037608</td>\n",
       "      <td>0.328014</td>\n",
       "      <td>0.045826</td>\n",
       "      <td>0.981409</td>\n",
       "      <td>0.912412</td>\n",
       "      <td>0.336965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>YAL012W</td>\n",
       "      <td>ATGACTCTACAAGAATCTGATAAATTTGCTACCAAGGCCATTCATG...</td>\n",
       "      <td>1185</td>\n",
       "      <td>YAL012W</td>\n",
       "      <td>568.34</td>\n",
       "      <td>1107.00</td>\n",
       "      <td>35.033287</td>\n",
       "      <td>0.739465</td>\n",
       "      <td>0.217966</td>\n",
       "      <td>0.805438</td>\n",
       "      <td>...</td>\n",
       "      <td>1.606033</td>\n",
       "      <td>1.198071</td>\n",
       "      <td>1.631042</td>\n",
       "      <td>0.556928</td>\n",
       "      <td>0.079883</td>\n",
       "      <td>0.147774</td>\n",
       "      <td>0.023571</td>\n",
       "      <td>2.242691</td>\n",
       "      <td>0.919870</td>\n",
       "      <td>0.259401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>YAL013W</td>\n",
       "      <td>ATGAGTCAGCAAACACCACAGGAAAGTGAACAGACCACAGCGAAAG...</td>\n",
       "      <td>1218</td>\n",
       "      <td>YAL013W</td>\n",
       "      <td>25.52</td>\n",
       "      <td>173.00</td>\n",
       "      <td>50.720205</td>\n",
       "      <td>0.593788</td>\n",
       "      <td>0.250821</td>\n",
       "      <td>0.700856</td>\n",
       "      <td>...</td>\n",
       "      <td>2.994454</td>\n",
       "      <td>1.093083</td>\n",
       "      <td>0.556058</td>\n",
       "      <td>0.651978</td>\n",
       "      <td>0.140033</td>\n",
       "      <td>0.277588</td>\n",
       "      <td>0.058232</td>\n",
       "      <td>3.019052</td>\n",
       "      <td>0.608649</td>\n",
       "      <td>0.213474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>YAL014C</td>\n",
       "      <td>ATGGATGTGTTGAAGCTGGGTTATGAACTGGACCAGCTCTCTGATC...</td>\n",
       "      <td>768</td>\n",
       "      <td>YAL014C</td>\n",
       "      <td>63.30</td>\n",
       "      <td>29.20</td>\n",
       "      <td>57.049019</td>\n",
       "      <td>0.655414</td>\n",
       "      <td>0.266603</td>\n",
       "      <td>0.646003</td>\n",
       "      <td>...</td>\n",
       "      <td>3.321965</td>\n",
       "      <td>1.184762</td>\n",
       "      <td>0.807917</td>\n",
       "      <td>0.695920</td>\n",
       "      <td>0.129407</td>\n",
       "      <td>0.243268</td>\n",
       "      <td>0.038314</td>\n",
       "      <td>3.130600</td>\n",
       "      <td>0.846709</td>\n",
       "      <td>0.304618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>YAL015C</td>\n",
       "      <td>ATGCAAAAGATCAGTAAATACTCATCTATGGCAATTCTGAGGAAAA...</td>\n",
       "      <td>1200</td>\n",
       "      <td>YAL015C</td>\n",
       "      <td>23.07</td>\n",
       "      <td>12.20</td>\n",
       "      <td>57.056301</td>\n",
       "      <td>0.693492</td>\n",
       "      <td>0.253924</td>\n",
       "      <td>0.684424</td>\n",
       "      <td>...</td>\n",
       "      <td>2.116035</td>\n",
       "      <td>1.070425</td>\n",
       "      <td>0.699774</td>\n",
       "      <td>0.592347</td>\n",
       "      <td>0.142825</td>\n",
       "      <td>0.333514</td>\n",
       "      <td>0.039014</td>\n",
       "      <td>1.827668</td>\n",
       "      <td>1.166366</td>\n",
       "      <td>0.532532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>YAL016C-A</td>\n",
       "      <td>ATGGTATCTAGTTTTTTCATGGCTTCTACCCTATTAGCAATATCAT...</td>\n",
       "      <td>315</td>\n",
       "      <td>YAL016C-A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61.000000</td>\n",
       "      <td>0.713958</td>\n",
       "      <td>0.237226</td>\n",
       "      <td>0.738951</td>\n",
       "      <td>...</td>\n",
       "      <td>0.839734</td>\n",
       "      <td>1.039142</td>\n",
       "      <td>0.923160</td>\n",
       "      <td>0.736868</td>\n",
       "      <td>0.239996</td>\n",
       "      <td>0.288589</td>\n",
       "      <td>0.030991</td>\n",
       "      <td>1.216507</td>\n",
       "      <td>1.171302</td>\n",
       "      <td>0.968614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>YAL016C-B</td>\n",
       "      <td>ATGCTGTATTTTTATCATATTAACATTTTGATTAAAGAAAGAAAGT...</td>\n",
       "      <td>186</td>\n",
       "      <td>YAL016C-B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.993904</td>\n",
       "      <td>0.838210</td>\n",
       "      <td>0.201266</td>\n",
       "      <td>0.830809</td>\n",
       "      <td>...</td>\n",
       "      <td>2.060356</td>\n",
       "      <td>1.079651</td>\n",
       "      <td>0.756554</td>\n",
       "      <td>0.796373</td>\n",
       "      <td>0.238221</td>\n",
       "      <td>0.324012</td>\n",
       "      <td>0.030065</td>\n",
       "      <td>2.337342</td>\n",
       "      <td>1.107911</td>\n",
       "      <td>0.628663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>YAL016W</td>\n",
       "      <td>ATGTCTGGAGCAAGATCAACAACGGCAGGTGCCGTGCCCTCGGCAG...</td>\n",
       "      <td>1908</td>\n",
       "      <td>YAL016W</td>\n",
       "      <td>143.30</td>\n",
       "      <td>294.00</td>\n",
       "      <td>46.079534</td>\n",
       "      <td>0.774087</td>\n",
       "      <td>0.230220</td>\n",
       "      <td>0.741633</td>\n",
       "      <td>...</td>\n",
       "      <td>2.050556</td>\n",
       "      <td>1.134601</td>\n",
       "      <td>0.480577</td>\n",
       "      <td>0.657244</td>\n",
       "      <td>0.150857</td>\n",
       "      <td>0.252471</td>\n",
       "      <td>0.038507</td>\n",
       "      <td>2.437245</td>\n",
       "      <td>0.866705</td>\n",
       "      <td>0.463491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>YAL017W</td>\n",
       "      <td>ATGCCCTACATCGGTGCTTCCAACCTCTCAGAACATTCATTTGTTA...</td>\n",
       "      <td>4071</td>\n",
       "      <td>YAL017W</td>\n",
       "      <td>50.59</td>\n",
       "      <td>39.50</td>\n",
       "      <td>52.761520</td>\n",
       "      <td>0.733797</td>\n",
       "      <td>0.247235</td>\n",
       "      <td>0.741228</td>\n",
       "      <td>...</td>\n",
       "      <td>1.847055</td>\n",
       "      <td>1.131061</td>\n",
       "      <td>0.527939</td>\n",
       "      <td>0.588844</td>\n",
       "      <td>0.212734</td>\n",
       "      <td>0.202496</td>\n",
       "      <td>0.027076</td>\n",
       "      <td>2.207949</td>\n",
       "      <td>0.956942</td>\n",
       "      <td>0.596528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>YAL018C</td>\n",
       "      <td>ATGAGTTTTACAGGTTCATTAGCGCTGGCAGGTATTGGTGGTTTGG...</td>\n",
       "      <td>978</td>\n",
       "      <td>YAL018C</td>\n",
       "      <td>0.15</td>\n",
       "      <td>1.57</td>\n",
       "      <td>50.678867</td>\n",
       "      <td>0.745524</td>\n",
       "      <td>0.237532</td>\n",
       "      <td>0.722040</td>\n",
       "      <td>...</td>\n",
       "      <td>2.151968</td>\n",
       "      <td>1.161307</td>\n",
       "      <td>0.886555</td>\n",
       "      <td>0.762283</td>\n",
       "      <td>0.237595</td>\n",
       "      <td>0.301584</td>\n",
       "      <td>0.030513</td>\n",
       "      <td>1.890367</td>\n",
       "      <td>1.090977</td>\n",
       "      <td>0.571839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>YAL019W</td>\n",
       "      <td>ATGAGTGGTTCGCATTCAAATGATGAGGATGACGTAGTGCAAGTGC...</td>\n",
       "      <td>3396</td>\n",
       "      <td>YAL019W</td>\n",
       "      <td>74.28</td>\n",
       "      <td>39.80</td>\n",
       "      <td>51.898143</td>\n",
       "      <td>0.734003</td>\n",
       "      <td>0.242176</td>\n",
       "      <td>0.655033</td>\n",
       "      <td>...</td>\n",
       "      <td>1.829631</td>\n",
       "      <td>1.053937</td>\n",
       "      <td>0.348639</td>\n",
       "      <td>0.523350</td>\n",
       "      <td>0.129404</td>\n",
       "      <td>0.295527</td>\n",
       "      <td>0.050221</td>\n",
       "      <td>1.999402</td>\n",
       "      <td>0.931989</td>\n",
       "      <td>0.548964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>YAL019W-A</td>\n",
       "      <td>ATGTTGCTGAGTGAACTCGTAGCAACCGCCTCCTCTCTGCCATACA...</td>\n",
       "      <td>570</td>\n",
       "      <td>YAL019W-A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>236.00</td>\n",
       "      <td>61.000000</td>\n",
       "      <td>0.605645</td>\n",
       "      <td>0.256052</td>\n",
       "      <td>0.534829</td>\n",
       "      <td>...</td>\n",
       "      <td>1.959838</td>\n",
       "      <td>0.916357</td>\n",
       "      <td>0.410742</td>\n",
       "      <td>0.532054</td>\n",
       "      <td>0.120625</td>\n",
       "      <td>0.267245</td>\n",
       "      <td>0.049452</td>\n",
       "      <td>2.028495</td>\n",
       "      <td>0.960021</td>\n",
       "      <td>0.294206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>YAL020C</td>\n",
       "      <td>ATGAGTTGTGTGTATGCGTTTGGGTCTAATGGGCAAAGGCAACTGG...</td>\n",
       "      <td>1002</td>\n",
       "      <td>YAL020C</td>\n",
       "      <td>25.42</td>\n",
       "      <td>4.03</td>\n",
       "      <td>54.963289</td>\n",
       "      <td>0.582908</td>\n",
       "      <td>0.260938</td>\n",
       "      <td>0.612051</td>\n",
       "      <td>...</td>\n",
       "      <td>1.799313</td>\n",
       "      <td>0.894436</td>\n",
       "      <td>0.269825</td>\n",
       "      <td>0.547030</td>\n",
       "      <td>0.085649</td>\n",
       "      <td>0.266260</td>\n",
       "      <td>0.074037</td>\n",
       "      <td>2.358922</td>\n",
       "      <td>0.530632</td>\n",
       "      <td>0.224393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>YAL021C</td>\n",
       "      <td>ATGAACGACCCTTCTTTACTAGGCTACCCTAATGTGGGCCCGCAGC...</td>\n",
       "      <td>2514</td>\n",
       "      <td>YAL021C</td>\n",
       "      <td>92.55</td>\n",
       "      <td>184.00</td>\n",
       "      <td>57.145329</td>\n",
       "      <td>0.673644</td>\n",
       "      <td>0.244628</td>\n",
       "      <td>0.584722</td>\n",
       "      <td>...</td>\n",
       "      <td>2.182919</td>\n",
       "      <td>0.995275</td>\n",
       "      <td>0.491138</td>\n",
       "      <td>0.537422</td>\n",
       "      <td>0.112064</td>\n",
       "      <td>0.218321</td>\n",
       "      <td>0.047904</td>\n",
       "      <td>2.288731</td>\n",
       "      <td>0.822692</td>\n",
       "      <td>0.243277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>YAL022C</td>\n",
       "      <td>ATGAGTACTAGTGCGGACACTGATACCATCAAGAAGCCAATCCTTG...</td>\n",
       "      <td>1554</td>\n",
       "      <td>YAL022C</td>\n",
       "      <td>61.02</td>\n",
       "      <td>4.86</td>\n",
       "      <td>56.460279</td>\n",
       "      <td>0.641914</td>\n",
       "      <td>0.244464</td>\n",
       "      <td>0.657897</td>\n",
       "      <td>...</td>\n",
       "      <td>1.752372</td>\n",
       "      <td>1.183435</td>\n",
       "      <td>0.691842</td>\n",
       "      <td>0.611821</td>\n",
       "      <td>0.179666</td>\n",
       "      <td>0.389165</td>\n",
       "      <td>0.038936</td>\n",
       "      <td>0.875472</td>\n",
       "      <td>1.036388</td>\n",
       "      <td>0.537005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>YAL023C</td>\n",
       "      <td>ATGTCCTCGTCTTCGTCTACCGGGTACAGCAAAAACAATGCCGCCC...</td>\n",
       "      <td>2280</td>\n",
       "      <td>YAL023C</td>\n",
       "      <td>342.36</td>\n",
       "      <td>131.00</td>\n",
       "      <td>45.834990</td>\n",
       "      <td>0.729541</td>\n",
       "      <td>0.235011</td>\n",
       "      <td>0.651771</td>\n",
       "      <td>...</td>\n",
       "      <td>2.801561</td>\n",
       "      <td>1.078057</td>\n",
       "      <td>0.420676</td>\n",
       "      <td>0.443474</td>\n",
       "      <td>0.066450</td>\n",
       "      <td>0.095370</td>\n",
       "      <td>0.018122</td>\n",
       "      <td>1.931404</td>\n",
       "      <td>0.886997</td>\n",
       "      <td>0.279528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>YAL024C</td>\n",
       "      <td>ATGGAAATATTTAGCCAGAAAGATTACTACCCGACTCCATCCTCAA...</td>\n",
       "      <td>4308</td>\n",
       "      <td>YAL024C</td>\n",
       "      <td>25.86</td>\n",
       "      <td>8.22</td>\n",
       "      <td>51.523083</td>\n",
       "      <td>0.735977</td>\n",
       "      <td>0.241484</td>\n",
       "      <td>0.713782</td>\n",
       "      <td>...</td>\n",
       "      <td>2.151968</td>\n",
       "      <td>1.055765</td>\n",
       "      <td>0.539413</td>\n",
       "      <td>0.580625</td>\n",
       "      <td>0.147469</td>\n",
       "      <td>0.242690</td>\n",
       "      <td>0.049230</td>\n",
       "      <td>2.151968</td>\n",
       "      <td>0.682753</td>\n",
       "      <td>0.385228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>YAL025C</td>\n",
       "      <td>ATGTCCGACGAAATTGTTTGGCAAGTGATTAATCAAAGTTTCTGCT...</td>\n",
       "      <td>921</td>\n",
       "      <td>YAL025C</td>\n",
       "      <td>58.06</td>\n",
       "      <td>118.00</td>\n",
       "      <td>49.147454</td>\n",
       "      <td>0.744792</td>\n",
       "      <td>0.233634</td>\n",
       "      <td>0.785407</td>\n",
       "      <td>...</td>\n",
       "      <td>2.650083</td>\n",
       "      <td>1.091394</td>\n",
       "      <td>0.688251</td>\n",
       "      <td>0.775779</td>\n",
       "      <td>0.313616</td>\n",
       "      <td>0.298002</td>\n",
       "      <td>0.030502</td>\n",
       "      <td>2.787776</td>\n",
       "      <td>1.116353</td>\n",
       "      <td>0.805607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>YAL026C</td>\n",
       "      <td>ATGAATGACGACAGAGAAACCCCCCCAAAGAGGAAACCTGGGGAGG...</td>\n",
       "      <td>4068</td>\n",
       "      <td>YAL026C</td>\n",
       "      <td>74.53</td>\n",
       "      <td>35.60</td>\n",
       "      <td>46.882148</td>\n",
       "      <td>0.773361</td>\n",
       "      <td>0.233940</td>\n",
       "      <td>0.656811</td>\n",
       "      <td>...</td>\n",
       "      <td>2.906608</td>\n",
       "      <td>0.940171</td>\n",
       "      <td>0.464322</td>\n",
       "      <td>0.566966</td>\n",
       "      <td>0.149146</td>\n",
       "      <td>0.115266</td>\n",
       "      <td>0.018361</td>\n",
       "      <td>2.582375</td>\n",
       "      <td>0.879315</td>\n",
       "      <td>0.422841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>YAL026C-A</td>\n",
       "      <td>ATGATACTACTCAAAAGAGAGGTAAGTACGGTGAATTACAAGATGC...</td>\n",
       "      <td>438</td>\n",
       "      <td>YAL026C-A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50.477569</td>\n",
       "      <td>0.685227</td>\n",
       "      <td>0.259347</td>\n",
       "      <td>0.614135</td>\n",
       "      <td>...</td>\n",
       "      <td>2.143114</td>\n",
       "      <td>1.086816</td>\n",
       "      <td>0.692999</td>\n",
       "      <td>0.809276</td>\n",
       "      <td>0.386835</td>\n",
       "      <td>0.375485</td>\n",
       "      <td>0.071209</td>\n",
       "      <td>1.931404</td>\n",
       "      <td>0.651333</td>\n",
       "      <td>0.448401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>YAL027W</td>\n",
       "      <td>ATGGCACCAAGTATAGCAACGGTAAAGATAGCCAGGGACATGGTTT...</td>\n",
       "      <td>786</td>\n",
       "      <td>YAL027W</td>\n",
       "      <td>18.86</td>\n",
       "      <td>35.00</td>\n",
       "      <td>56.862760</td>\n",
       "      <td>0.717825</td>\n",
       "      <td>0.239508</td>\n",
       "      <td>0.730469</td>\n",
       "      <td>...</td>\n",
       "      <td>2.230956</td>\n",
       "      <td>1.101257</td>\n",
       "      <td>0.525975</td>\n",
       "      <td>0.700589</td>\n",
       "      <td>0.174015</td>\n",
       "      <td>0.166185</td>\n",
       "      <td>0.020275</td>\n",
       "      <td>2.338365</td>\n",
       "      <td>0.967295</td>\n",
       "      <td>0.530906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6681</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>ATATTAATAATATATATATTATTTATTATAATGAAAACCTATCCTA...</td>\n",
       "      <td>291</td>\n",
       "      <td>Q0032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.853662</td>\n",
       "      <td>0.188169</td>\n",
       "      <td>0.812326</td>\n",
       "      <td>...</td>\n",
       "      <td>1.939662</td>\n",
       "      <td>0.877433</td>\n",
       "      <td>0.316553</td>\n",
       "      <td>0.792972</td>\n",
       "      <td>0.188022</td>\n",
       "      <td>0.214752</td>\n",
       "      <td>0.032499</td>\n",
       "      <td>1.554128</td>\n",
       "      <td>0.632337</td>\n",
       "      <td>0.252417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6682</th>\n",
       "      <td>Q0045</td>\n",
       "      <td>ATGGTACAAAGATGATTATATTCAACAAATGCAAAAGATATTGCAG...</td>\n",
       "      <td>1605</td>\n",
       "      <td>Q0045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.860759</td>\n",
       "      <td>0.175470</td>\n",
       "      <td>0.904945</td>\n",
       "      <td>...</td>\n",
       "      <td>2.009395</td>\n",
       "      <td>1.197498</td>\n",
       "      <td>0.948229</td>\n",
       "      <td>0.749534</td>\n",
       "      <td>0.272762</td>\n",
       "      <td>0.304134</td>\n",
       "      <td>0.053197</td>\n",
       "      <td>2.047507</td>\n",
       "      <td>0.554785</td>\n",
       "      <td>0.397785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683</th>\n",
       "      <td>Q0050</td>\n",
       "      <td>ATGGTACAAAGATGATTATATTCAACAAATGCAAAAGATATTGCAG...</td>\n",
       "      <td>2505</td>\n",
       "      <td>Q0050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.859441</td>\n",
       "      <td>0.186742</td>\n",
       "      <td>0.904945</td>\n",
       "      <td>...</td>\n",
       "      <td>2.009395</td>\n",
       "      <td>1.197498</td>\n",
       "      <td>0.948229</td>\n",
       "      <td>0.749534</td>\n",
       "      <td>0.272762</td>\n",
       "      <td>0.304134</td>\n",
       "      <td>0.053197</td>\n",
       "      <td>2.047507</td>\n",
       "      <td>0.554785</td>\n",
       "      <td>0.397785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6684</th>\n",
       "      <td>Q0055</td>\n",
       "      <td>ATGGTACAAAGATGATTATATTCAACAAATGCAAAAGATATTGCAG...</td>\n",
       "      <td>2565</td>\n",
       "      <td>Q0055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.864181</td>\n",
       "      <td>0.190262</td>\n",
       "      <td>0.904945</td>\n",
       "      <td>...</td>\n",
       "      <td>2.009395</td>\n",
       "      <td>1.197498</td>\n",
       "      <td>0.948229</td>\n",
       "      <td>0.749534</td>\n",
       "      <td>0.272762</td>\n",
       "      <td>0.304134</td>\n",
       "      <td>0.053197</td>\n",
       "      <td>2.047507</td>\n",
       "      <td>0.554785</td>\n",
       "      <td>0.397785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6685</th>\n",
       "      <td>Q0060</td>\n",
       "      <td>ATGGTACAAAGATGATTATATTCAACAAATGCAAAAGATATTGCAG...</td>\n",
       "      <td>1248</td>\n",
       "      <td>Q0060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.893661</td>\n",
       "      <td>0.169984</td>\n",
       "      <td>0.904945</td>\n",
       "      <td>...</td>\n",
       "      <td>2.009395</td>\n",
       "      <td>1.197498</td>\n",
       "      <td>0.948229</td>\n",
       "      <td>0.749534</td>\n",
       "      <td>0.272762</td>\n",
       "      <td>0.304134</td>\n",
       "      <td>0.053197</td>\n",
       "      <td>2.047507</td>\n",
       "      <td>0.554785</td>\n",
       "      <td>0.397785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6686</th>\n",
       "      <td>Q0065</td>\n",
       "      <td>ATGGTACAAAGATGATTATATTCAACAAATGCAAAAGATATTGCAG...</td>\n",
       "      <td>1671</td>\n",
       "      <td>Q0065</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.887699</td>\n",
       "      <td>0.165679</td>\n",
       "      <td>0.904945</td>\n",
       "      <td>...</td>\n",
       "      <td>2.009395</td>\n",
       "      <td>1.197498</td>\n",
       "      <td>0.948229</td>\n",
       "      <td>0.749534</td>\n",
       "      <td>0.272762</td>\n",
       "      <td>0.304134</td>\n",
       "      <td>0.053197</td>\n",
       "      <td>2.047507</td>\n",
       "      <td>0.554785</td>\n",
       "      <td>0.397785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6687</th>\n",
       "      <td>Q0070</td>\n",
       "      <td>ATGGTACAAAGATGATTATATTCAACAAATGCAAAAGATATTGCAG...</td>\n",
       "      <td>1893</td>\n",
       "      <td>Q0070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.883968</td>\n",
       "      <td>0.166908</td>\n",
       "      <td>0.904945</td>\n",
       "      <td>...</td>\n",
       "      <td>2.009395</td>\n",
       "      <td>1.197498</td>\n",
       "      <td>0.948229</td>\n",
       "      <td>0.749534</td>\n",
       "      <td>0.272762</td>\n",
       "      <td>0.304134</td>\n",
       "      <td>0.053197</td>\n",
       "      <td>2.047507</td>\n",
       "      <td>0.554785</td>\n",
       "      <td>0.397785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6688</th>\n",
       "      <td>Q0075</td>\n",
       "      <td>ATATTAATATTATTAATAATAATTTTAACTAATAATAAAGTTTTTA...</td>\n",
       "      <td>1065</td>\n",
       "      <td>Q0075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.921349</td>\n",
       "      <td>0.151766</td>\n",
       "      <td>0.818235</td>\n",
       "      <td>...</td>\n",
       "      <td>1.645678</td>\n",
       "      <td>1.024213</td>\n",
       "      <td>0.491549</td>\n",
       "      <td>0.468005</td>\n",
       "      <td>0.087481</td>\n",
       "      <td>0.254259</td>\n",
       "      <td>0.044276</td>\n",
       "      <td>2.165634</td>\n",
       "      <td>0.971329</td>\n",
       "      <td>0.374636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6689</th>\n",
       "      <td>Q0080</td>\n",
       "      <td>ATGCCACAATTAGTTCCATTTTATTTTATGAATCAATTAACATATG...</td>\n",
       "      <td>147</td>\n",
       "      <td>Q0080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49.247059</td>\n",
       "      <td>0.922732</td>\n",
       "      <td>0.138107</td>\n",
       "      <td>0.922732</td>\n",
       "      <td>...</td>\n",
       "      <td>1.232419</td>\n",
       "      <td>0.968659</td>\n",
       "      <td>0.855447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6690</th>\n",
       "      <td>Q0085</td>\n",
       "      <td>ATGTTTAATTTATTAAATACATATATTACATCACCATTAGATCAAT...</td>\n",
       "      <td>780</td>\n",
       "      <td>Q0085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.901373</td>\n",
       "      <td>0.154858</td>\n",
       "      <td>0.925518</td>\n",
       "      <td>...</td>\n",
       "      <td>0.589680</td>\n",
       "      <td>1.034387</td>\n",
       "      <td>0.473624</td>\n",
       "      <td>0.914116</td>\n",
       "      <td>0.251373</td>\n",
       "      <td>0.287413</td>\n",
       "      <td>0.046083</td>\n",
       "      <td>0.299755</td>\n",
       "      <td>0.565358</td>\n",
       "      <td>0.407886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6691</th>\n",
       "      <td>Q0092</td>\n",
       "      <td>ATATTTATAATAAATAATAATATATTCATTATATTTATAATTATAT...</td>\n",
       "      <td>141</td>\n",
       "      <td>Q0092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38.571429</td>\n",
       "      <td>0.703248</td>\n",
       "      <td>0.252711</td>\n",
       "      <td>0.703248</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6692</th>\n",
       "      <td>Q0105</td>\n",
       "      <td>ATGGCATTTAGAAAATCAAATGTGTATTTAAGTTTAGTGAATAGTT...</td>\n",
       "      <td>1158</td>\n",
       "      <td>Q0105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.875898</td>\n",
       "      <td>0.169834</td>\n",
       "      <td>0.915731</td>\n",
       "      <td>...</td>\n",
       "      <td>1.290364</td>\n",
       "      <td>1.144637</td>\n",
       "      <td>0.825314</td>\n",
       "      <td>0.875655</td>\n",
       "      <td>0.269225</td>\n",
       "      <td>0.339034</td>\n",
       "      <td>0.055033</td>\n",
       "      <td>0.749146</td>\n",
       "      <td>0.564957</td>\n",
       "      <td>0.400131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6693</th>\n",
       "      <td>Q0110</td>\n",
       "      <td>ATGGCATTTAGAAAATCAAATGTGTATTTAAGTTTAGTGAATAGTT...</td>\n",
       "      <td>1272</td>\n",
       "      <td>Q0110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.906642</td>\n",
       "      <td>0.153883</td>\n",
       "      <td>0.915731</td>\n",
       "      <td>...</td>\n",
       "      <td>1.290364</td>\n",
       "      <td>1.144637</td>\n",
       "      <td>0.825314</td>\n",
       "      <td>0.875655</td>\n",
       "      <td>0.269225</td>\n",
       "      <td>0.339034</td>\n",
       "      <td>0.055033</td>\n",
       "      <td>0.749146</td>\n",
       "      <td>0.564957</td>\n",
       "      <td>0.400131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6694</th>\n",
       "      <td>Q0115</td>\n",
       "      <td>ATGGCATTTAGAAAATCAAATGTGTATTTAAGTTTAGTGAATAGTT...</td>\n",
       "      <td>1554</td>\n",
       "      <td>Q0115</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.905015</td>\n",
       "      <td>0.156940</td>\n",
       "      <td>0.915731</td>\n",
       "      <td>...</td>\n",
       "      <td>1.290364</td>\n",
       "      <td>1.144637</td>\n",
       "      <td>0.825314</td>\n",
       "      <td>0.875655</td>\n",
       "      <td>0.269225</td>\n",
       "      <td>0.339034</td>\n",
       "      <td>0.055033</td>\n",
       "      <td>0.749146</td>\n",
       "      <td>0.564957</td>\n",
       "      <td>0.400131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6695</th>\n",
       "      <td>Q0120</td>\n",
       "      <td>ATGGCATTTAGAAAATCAAATGTGTATTTAAGTTTAGTGAATAGTT...</td>\n",
       "      <td>1917</td>\n",
       "      <td>Q0120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.899115</td>\n",
       "      <td>0.165019</td>\n",
       "      <td>0.915731</td>\n",
       "      <td>...</td>\n",
       "      <td>1.290364</td>\n",
       "      <td>1.144637</td>\n",
       "      <td>0.825314</td>\n",
       "      <td>0.875655</td>\n",
       "      <td>0.269225</td>\n",
       "      <td>0.339034</td>\n",
       "      <td>0.055033</td>\n",
       "      <td>0.749146</td>\n",
       "      <td>0.564957</td>\n",
       "      <td>0.400131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6696</th>\n",
       "      <td>Q0130</td>\n",
       "      <td>ATGCAATTAGTATTAGCAGCTAAATATATTGGAGCAGGTATCTCAA...</td>\n",
       "      <td>231</td>\n",
       "      <td>Q0130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29.761364</td>\n",
       "      <td>0.846472</td>\n",
       "      <td>0.174538</td>\n",
       "      <td>0.834434</td>\n",
       "      <td>...</td>\n",
       "      <td>2.235826</td>\n",
       "      <td>1.099273</td>\n",
       "      <td>0.697846</td>\n",
       "      <td>0.797130</td>\n",
       "      <td>0.361863</td>\n",
       "      <td>0.385061</td>\n",
       "      <td>0.040555</td>\n",
       "      <td>2.528694</td>\n",
       "      <td>0.891943</td>\n",
       "      <td>0.744431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6697</th>\n",
       "      <td>Q0140</td>\n",
       "      <td>ATGAAATTAAAATTATTAAATATAATTTTATCAATAATAAATAAAC...</td>\n",
       "      <td>1197</td>\n",
       "      <td>Q0140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.903100</td>\n",
       "      <td>0.167229</td>\n",
       "      <td>0.888527</td>\n",
       "      <td>...</td>\n",
       "      <td>1.078655</td>\n",
       "      <td>1.150314</td>\n",
       "      <td>0.571484</td>\n",
       "      <td>0.861310</td>\n",
       "      <td>0.149759</td>\n",
       "      <td>0.246185</td>\n",
       "      <td>0.041690</td>\n",
       "      <td>0.815903</td>\n",
       "      <td>0.499392</td>\n",
       "      <td>0.228332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6698</th>\n",
       "      <td>Q0142</td>\n",
       "      <td>ATGCTGGGTTCCGGAACTCCTCCTTCTCGCGAGGTTAACACCTATT...</td>\n",
       "      <td>177</td>\n",
       "      <td>Q0142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48.941547</td>\n",
       "      <td>0.774572</td>\n",
       "      <td>0.233263</td>\n",
       "      <td>0.762008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.255881</td>\n",
       "      <td>0.936553</td>\n",
       "      <td>0.236193</td>\n",
       "      <td>0.675520</td>\n",
       "      <td>0.107744</td>\n",
       "      <td>0.207140</td>\n",
       "      <td>0.030349</td>\n",
       "      <td>0.442457</td>\n",
       "      <td>0.673926</td>\n",
       "      <td>0.193477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6699</th>\n",
       "      <td>Q0143</td>\n",
       "      <td>ATGGGGTTATGGATTTCGTTCGGAACTCCTCCCTCCTACCTCTATT...</td>\n",
       "      <td>153</td>\n",
       "      <td>Q0143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>0.250650</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>...</td>\n",
       "      <td>0.478332</td>\n",
       "      <td>0.926792</td>\n",
       "      <td>0.417405</td>\n",
       "      <td>0.780918</td>\n",
       "      <td>0.172752</td>\n",
       "      <td>0.268658</td>\n",
       "      <td>0.027006</td>\n",
       "      <td>1.046744</td>\n",
       "      <td>1.020796</td>\n",
       "      <td>0.384229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6700</th>\n",
       "      <td>Q0144</td>\n",
       "      <td>ATATTCATATTTTTTATTAATATTATTTTTATTAATAACTTTTTAA...</td>\n",
       "      <td>330</td>\n",
       "      <td>Q0144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.858007</td>\n",
       "      <td>0.189924</td>\n",
       "      <td>0.884038</td>\n",
       "      <td>...</td>\n",
       "      <td>0.768998</td>\n",
       "      <td>0.908497</td>\n",
       "      <td>0.332013</td>\n",
       "      <td>0.880132</td>\n",
       "      <td>0.213358</td>\n",
       "      <td>0.327436</td>\n",
       "      <td>0.052921</td>\n",
       "      <td>0.827805</td>\n",
       "      <td>0.572753</td>\n",
       "      <td>0.315722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6701</th>\n",
       "      <td>Q0160</td>\n",
       "      <td>ATGAAAAATATTAAAAAAAATCAAGTAATAAATTTAGGACCTAATT...</td>\n",
       "      <td>708</td>\n",
       "      <td>Q0160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.905549</td>\n",
       "      <td>0.159184</td>\n",
       "      <td>0.887633</td>\n",
       "      <td>...</td>\n",
       "      <td>1.140365</td>\n",
       "      <td>1.121351</td>\n",
       "      <td>0.758280</td>\n",
       "      <td>0.662386</td>\n",
       "      <td>0.216106</td>\n",
       "      <td>0.315021</td>\n",
       "      <td>0.042994</td>\n",
       "      <td>0.838458</td>\n",
       "      <td>0.693256</td>\n",
       "      <td>0.491948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6702</th>\n",
       "      <td>Q0182</td>\n",
       "      <td>ATAGTTTCGGGGCCCGGCCACGGGAGTCGGAACCCCGAAAGGAGTT...</td>\n",
       "      <td>405</td>\n",
       "      <td>Q0182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.718684</td>\n",
       "      <td>0.270808</td>\n",
       "      <td>0.610440</td>\n",
       "      <td>...</td>\n",
       "      <td>3.196348</td>\n",
       "      <td>0.896042</td>\n",
       "      <td>0.247977</td>\n",
       "      <td>0.608300</td>\n",
       "      <td>0.098185</td>\n",
       "      <td>0.303016</td>\n",
       "      <td>0.082693</td>\n",
       "      <td>2.724260</td>\n",
       "      <td>0.465539</td>\n",
       "      <td>0.225940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6703</th>\n",
       "      <td>Q0250</td>\n",
       "      <td>ATGTTAGATTTATTAAGATTACAATTAACAACATTCATTATGAATG...</td>\n",
       "      <td>756</td>\n",
       "      <td>Q0250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.921855</td>\n",
       "      <td>0.138906</td>\n",
       "      <td>0.941335</td>\n",
       "      <td>...</td>\n",
       "      <td>1.070744</td>\n",
       "      <td>1.172725</td>\n",
       "      <td>0.846132</td>\n",
       "      <td>0.846820</td>\n",
       "      <td>0.241245</td>\n",
       "      <td>0.289552</td>\n",
       "      <td>0.044233</td>\n",
       "      <td>0.741281</td>\n",
       "      <td>0.642625</td>\n",
       "      <td>0.328154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6704</th>\n",
       "      <td>Q0255</td>\n",
       "      <td>ATGATTAAATGAACAATAATTAATATTTACTTATTATTAATATTTT...</td>\n",
       "      <td>1419</td>\n",
       "      <td>Q0255</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.918472</td>\n",
       "      <td>0.150875</td>\n",
       "      <td>0.931864</td>\n",
       "      <td>...</td>\n",
       "      <td>0.735499</td>\n",
       "      <td>1.078698</td>\n",
       "      <td>0.473959</td>\n",
       "      <td>0.732440</td>\n",
       "      <td>0.189256</td>\n",
       "      <td>0.263259</td>\n",
       "      <td>0.033273</td>\n",
       "      <td>0.735499</td>\n",
       "      <td>0.752935</td>\n",
       "      <td>0.341381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6705</th>\n",
       "      <td>Q0275</td>\n",
       "      <td>ATGACACATTTAGAAAGAAGTAGACATCAACAACATCCATTTCATA...</td>\n",
       "      <td>810</td>\n",
       "      <td>Q0275</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.881556</td>\n",
       "      <td>0.165733</td>\n",
       "      <td>0.890958</td>\n",
       "      <td>...</td>\n",
       "      <td>2.617123</td>\n",
       "      <td>1.081502</td>\n",
       "      <td>0.516189</td>\n",
       "      <td>0.773270</td>\n",
       "      <td>0.214411</td>\n",
       "      <td>0.266783</td>\n",
       "      <td>0.059646</td>\n",
       "      <td>2.023010</td>\n",
       "      <td>0.502121</td>\n",
       "      <td>0.278518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6706</th>\n",
       "      <td>Q0297</td>\n",
       "      <td>ATGAAAATAAAACTTAATAAATATATAAATATAGTCCGGCCCGCCC...</td>\n",
       "      <td>156</td>\n",
       "      <td>Q0297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61.000000</td>\n",
       "      <td>0.558868</td>\n",
       "      <td>0.289727</td>\n",
       "      <td>0.558868</td>\n",
       "      <td>...</td>\n",
       "      <td>2.065909</td>\n",
       "      <td>1.033023</td>\n",
       "      <td>0.323567</td>\n",
       "      <td>0.669068</td>\n",
       "      <td>0.155689</td>\n",
       "      <td>0.161217</td>\n",
       "      <td>0.022486</td>\n",
       "      <td>1.839058</td>\n",
       "      <td>0.748217</td>\n",
       "      <td>0.326807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6707</th>\n",
       "      <td>R0010W</td>\n",
       "      <td>ATGCCACAATTTGGTATATTATGTAAAACACCACCTAAGGTGCTTG...</td>\n",
       "      <td>1272</td>\n",
       "      <td>R0010W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>51.151812</td>\n",
       "      <td>0.727717</td>\n",
       "      <td>0.239465</td>\n",
       "      <td>0.791688</td>\n",
       "      <td>...</td>\n",
       "      <td>2.917612</td>\n",
       "      <td>1.006726</td>\n",
       "      <td>0.738298</td>\n",
       "      <td>0.660113</td>\n",
       "      <td>0.145917</td>\n",
       "      <td>0.248094</td>\n",
       "      <td>0.042628</td>\n",
       "      <td>2.547540</td>\n",
       "      <td>0.672295</td>\n",
       "      <td>0.312577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6708</th>\n",
       "      <td>R0020C</td>\n",
       "      <td>ATGAATGGCGAGAGACTGCTTGCTTGTATTAAGCAATGTATTATGC...</td>\n",
       "      <td>1122</td>\n",
       "      <td>R0020C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>55.816919</td>\n",
       "      <td>0.705527</td>\n",
       "      <td>0.249949</td>\n",
       "      <td>0.728294</td>\n",
       "      <td>...</td>\n",
       "      <td>2.316916</td>\n",
       "      <td>1.039254</td>\n",
       "      <td>0.538441</td>\n",
       "      <td>0.624505</td>\n",
       "      <td>0.123525</td>\n",
       "      <td>0.248607</td>\n",
       "      <td>0.039779</td>\n",
       "      <td>2.760384</td>\n",
       "      <td>0.829067</td>\n",
       "      <td>0.329799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6709</th>\n",
       "      <td>R0030W</td>\n",
       "      <td>ATGCCTTATAAAACAGCTATAGATTGCATAGAAGAGTTAGCTACTC...</td>\n",
       "      <td>546</td>\n",
       "      <td>R0030W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.824413</td>\n",
       "      <td>0.674440</td>\n",
       "      <td>0.270264</td>\n",
       "      <td>0.708686</td>\n",
       "      <td>...</td>\n",
       "      <td>2.207949</td>\n",
       "      <td>0.996945</td>\n",
       "      <td>0.515097</td>\n",
       "      <td>0.675505</td>\n",
       "      <td>0.188087</td>\n",
       "      <td>0.326574</td>\n",
       "      <td>0.045184</td>\n",
       "      <td>1.875012</td>\n",
       "      <td>0.857342</td>\n",
       "      <td>0.535492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6710</th>\n",
       "      <td>R0040C</td>\n",
       "      <td>ATGGACGACATTGAAACAGCCAAGAATCTGACGGTAAAAGCACGTA...</td>\n",
       "      <td>891</td>\n",
       "      <td>R0040C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52.165255</td>\n",
       "      <td>0.706931</td>\n",
       "      <td>0.257872</td>\n",
       "      <td>0.679196</td>\n",
       "      <td>...</td>\n",
       "      <td>2.308589</td>\n",
       "      <td>0.989668</td>\n",
       "      <td>0.432865</td>\n",
       "      <td>0.659205</td>\n",
       "      <td>0.169835</td>\n",
       "      <td>0.249536</td>\n",
       "      <td>0.044412</td>\n",
       "      <td>2.328964</td>\n",
       "      <td>0.662251</td>\n",
       "      <td>0.355036</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6711 rows × 732 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ORF                                      gene_sequence  length  \\\n",
       "0     YAL001C  ATGGTACTGACGATTTATCCTGACGAACTCGTACAAATAGTGTCTG...    3483   \n",
       "1     YAL002W  ATGGAGCAAAATGGCCTTGACCACGACAGCAGATCTAGCATCGATA...    3825   \n",
       "2     YAL003W  ATGGCATCCACCGATTTCTCCAAGATTGAAACTTTGAAACAATTAA...     621   \n",
       "3     YAL004W  ATGGGTGTCACCAGCGGTGGCCTTAACTTCAAAGATACCGTCTTCA...     648   \n",
       "4     YAL005C  ATGTCAAAAGCTGTCGGTATTGATTTAGGTACAACATACTCGTGTG...    1929   \n",
       "...       ...                                                ...     ...   \n",
       "6706    Q0297  ATGAAAATAAAACTTAATAAATATATAAATATAGTCCGGCCCGCCC...     156   \n",
       "6707   R0010W  ATGCCACAATTTGGTATATTATGTAAAACACCACCTAAGGTGCTTG...    1272   \n",
       "6708   R0020C  ATGAATGGCGAGAGACTGCTTGCTTGTATTAAGCAATGTATTATGC...    1122   \n",
       "6709   R0030W  ATGCCTTATAAAACAGCTATAGATTGCATAGAAGAGTTAGCTACTC...     546   \n",
       "6710   R0040C  ATGGACGACATTGAAACAGCCAAGAATCTGACGGTAAAAGCACGTA...     891   \n",
       "\n",
       "          ORF  mRNA_RPKM       PA        ENC  CAI_index_wholeGenome  \\\n",
       "0     YAL001C      37.80    11.60  52.499762               0.734169   \n",
       "1     YAL002W      22.17     7.27  50.928649               0.742818   \n",
       "2     YAL003W    4621.86  3029.00  26.669428               0.825756   \n",
       "3     YAL004W        NaN    26.80  34.435740               0.674925   \n",
       "4     YAL005C        NaN  1117.00  28.341978               0.823721   \n",
       "...       ...        ...      ...        ...                    ...   \n",
       "6706    Q0297        NaN      NaN  61.000000               0.558868   \n",
       "6707   R0010W        NaN      NaN  51.151812               0.727717   \n",
       "6708   R0020C        NaN      NaN  55.816919               0.705527   \n",
       "6709   R0030W        NaN      NaN  57.824413               0.674440   \n",
       "6710   R0040C        NaN      NaN  52.165255               0.706931   \n",
       "\n",
       "      CAI_STD_wholeGenome  CAI_first_50_codons_wholeGenome  ...  RCBS_win99  \\\n",
       "0                0.250486                         0.759264  ...    2.750944   \n",
       "1                0.239425                         0.714702  ...    3.061334   \n",
       "2                0.190030                         0.836821  ...    2.118197   \n",
       "3                0.240300                         0.659022  ...    3.178232   \n",
       "4                0.195552                         0.837872  ...    2.391503   \n",
       "...                   ...                              ...  ...         ...   \n",
       "6706             0.289727                         0.558868  ...    2.065909   \n",
       "6707             0.239465                         0.791688  ...    2.917612   \n",
       "6708             0.249949                         0.728294  ...    2.316916   \n",
       "6709             0.270264                         0.708686  ...    2.207949   \n",
       "6710             0.257872                         0.679196  ...    2.308589   \n",
       "\n",
       "      RCA_index_win99_wholeGenome  RCA_index_win99_HEG  \\\n",
       "0                        1.222763             0.787172   \n",
       "1                        1.092761             0.320760   \n",
       "2                        1.197576             1.908536   \n",
       "3                        0.991119             0.402804   \n",
       "4                        1.175160             1.734433   \n",
       "...                           ...                  ...   \n",
       "6706                     1.033023             0.323567   \n",
       "6707                     1.006726             0.738298   \n",
       "6708                     1.039254             0.538441   \n",
       "6709                     0.996945             0.515097   \n",
       "6710                     0.989668             0.432865   \n",
       "\n",
       "      CAI_index_win100_wholeGenome  CAI_index_win100_HEG  tAI_index_win100  \\\n",
       "0                         0.707072              0.178779          0.241931   \n",
       "1                         0.523649              0.093294          0.189421   \n",
       "2                         0.640120              0.117762          0.193914   \n",
       "3                         0.637584              0.123598          0.297283   \n",
       "4                         0.665937              0.123181          0.187328   \n",
       "...                            ...                   ...               ...   \n",
       "6706                      0.669068              0.155689          0.161217   \n",
       "6707                      0.660113              0.145917          0.248094   \n",
       "6708                      0.624505              0.123525          0.248607   \n",
       "6709                      0.675505              0.188087          0.326574   \n",
       "6710                      0.659205              0.169835          0.249536   \n",
       "\n",
       "      NTE_index_win100  RCBS_win100  RCA_index_win100_wholeGenome  \\\n",
       "0             0.027772     1.879886                      1.032415   \n",
       "1             0.046871     2.131717                      0.605164   \n",
       "2             0.025949     2.458412                      1.115684   \n",
       "3             0.043497     2.118197                      0.883690   \n",
       "4             0.027386     2.358922                      0.987152   \n",
       "...                ...          ...                           ...   \n",
       "6706          0.022486     1.839058                      0.748217   \n",
       "6707          0.042628     2.547540                      0.672295   \n",
       "6708          0.039779     2.760384                      0.829067   \n",
       "6709          0.045184     1.875012                      0.857342   \n",
       "6710          0.044412     2.328964                      0.662251   \n",
       "\n",
       "      RCA_index_win100_HEG  \n",
       "0                 0.405449  \n",
       "1                 0.262181  \n",
       "2                 0.371074  \n",
       "3                 0.389291  \n",
       "4                 0.399857  \n",
       "...                    ...  \n",
       "6706              0.326807  \n",
       "6707              0.312577  \n",
       "6708              0.329799  \n",
       "6709              0.535492  \n",
       "6710              0.355036  \n",
       "\n",
       "[6711 rows x 732 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
